---
title: "DGE analysis"
author: "Hugo Scharfenstein"
date: "`r Sys.Date()`"
output: html_document
---

Packages used for analysis
```{r}
#General packages
library(tidyverse)
library(readxl)
#DGE analysis packages
library(DESeq2)
library(edgeR)
library(glmpca)
library(apeglm)
library(ashr)
#Plotting packages
library(pheatmap)
library(EnhancedVolcano)
library(DEGreport)
library(RColorBrewer)
library(ggpubr)
library(plotly)
```

1. Import gene count matrices generated from STAR and associated metadata
```{r}
#Import metadata
metadata <- read_xlsx("/Users/hugo/Documents/Melbourne PhD/Experimental work/Chapter 5/Gene counts matrices/rnaseq_metadata.xlsx", sheet = 1)

#Import the gene count matrix resulting from the mapping of sequences to the C. proliferum reference genome
host_counts <- read_tsv("/Users/hugo/Documents/Melbourne PhD/Experimental work/Chapter 5/Gene counts matrices/gene_counts_matrix_pdae.txt")

#Import the gene count matrix resulting from the mapping of sequences to the C. proliferum reference genome
sym_counts <- read_tsv("/Users/hugo/Documents/Melbourne PhD/Experimental work/Chapter 5/Gene counts matrices/gene_counts_matrix_cprol.txt")
```

2. Collate data to gene level
```{r}
#Host
host_gene_counts <- host_counts %>% 
  separate(gene_id, c("gene", "exon"), sep = "\\.exon") %>% 
  group_by(gene) %>% 
  summarise_if(is.numeric, sum)

#Symbiont
sym_gene_counts <- sym_counts %>% 
  separate(gene_id, c("gene", "exon"), sep = "\\.exon") %>% 
  group_by(gene) %>% 
  summarise_if(is.numeric, sum)
```

3. Decide on appropriate statistical model for downstream analysis 
```{r}
#Visualise RNA-seq count distribution for one sample
ggplot(sym_gene_counts) + #Change input file to switch between host and symbiont samples
  geom_histogram(aes(x = Amb_C_I_1005_S22_), stat = "bin", bins = 200) +
  xlim(-5, 500) +
  xlab("Raw expression counts") +
  ylab("Number of genes")

#Calculate mean and variance of count data
mean_counts <- apply(sym_gene_counts[, 3:5], 1, mean) #Change input file 
variance_counts <- apply(sym_gene_counts[, 3:5], 1, var) #Change input file 
df <- data.frame(mean_counts, variance_counts)

#Plot mean versus variance
ggplot(df) +
        geom_point(aes(x=mean_counts, y=variance_counts)) + 
        geom_line(aes(x=mean_counts, y=mean_counts, color="red")) +
        scale_y_log10() +
        scale_x_log10()
#Overall, the variance appears to be greater than mean of counts
#This rules out use of Poisson distribution where mean=variance
#Rather, a negative binomial model is more appropriate to this analysis (in this case DESeq2 will be used)
```

4. Preparation of gene count matrices for downstream analysis

4.1. Load raw gene counts into DESeqDataSet objects 
```{r}
#Modify the metadata table, adding  sample names as row names (needed for some of the DESeq functions)
metadata <- as.data.frame(metadata)
rownames(metadata) <- metadata$SampleName

#Modify the host/sym_gene_counts tables, adding  gene ids as row names (needed for later)
host_gene_counts <- as.data.frame(host_gene_counts)
rownames(host_gene_counts) <- host_gene_counts$gene
host_gene_counts <- host_gene_counts[,-1]

sym_gene_counts <- as.data.frame(sym_gene_counts)
rownames(sym_gene_counts) <- sym_gene_counts$gene
sym_gene_counts <- sym_gene_counts[,-1]

#Check if row and column names match between metadata and host/sym_gene_counts
all(colnames(host_gene_counts) %in% rownames(metadata)) 
all(colnames(host_gene_counts) == rownames(metadata))

all(colnames(sym_gene_counts) %in% rownames(metadata)) 
all(colnames(sym_gene_counts) == rownames(metadata))

#Load the gene count matrices from STAR into a DESeq DESeqDataSet objects
host_dds_raw <- DESeqDataSetFromMatrix(countData = host_gene_counts,
                                       colData = metadata,
                                       design = ~  RI_treatment + Temp_Col) 

sym_dds_raw <- DESeqDataSetFromMatrix(countData = sym_gene_counts,
                                      colData = metadata,
                                      design = ~ Colony + Temperature + RI_treatment + Temperature:RI_treatment)
#For experimental design options check results section in: http://bioconductor.org/packages/3.17/bioc/manuals/DESeq2/man/DESeq2.pdf
```

4.2. Filter out lowly expressed genes (genes where normalised gene counts are >= 10 across at least 2 samples are kept)
```{r}
#Option 1: using CPM as a threshold (using function cpm from edgeR)
#Check following paper: https://doi.org/10.12688/f1000research.8987.2
host_cpm_filt <- rowSums(cpm(host_dds_raw) >= 0.28) >= 2 #A CPM of 0.25 should correspond to ~10 normalised gened counts (CPM is determined by: gene count threshold/minimum library size in millions, ~36M for the host)
#The threshold for the number of samples is set to the smallest group size (in this case n=2)
table(host_cpm_filt) #Check how many genes were filtered out
host_dds_filt1 <- host_dds_raw[host_cpm_filt,]

sym_cpm_filt <- rowSums(cpm(sym_dds_raw) >= 5) >= 2
table(sym_cpm_filt)
sym_dds_filt1 <- sym_dds_raw[sym_cpm_filt,]

#Option 2: perform the median of ratios method of normalization (using DESeq2)
host_dds_raw <- estimateSizeFactors(host_dds_raw)
host_mrm_filt <- rowSums(counts(host_dds_raw, normalized = TRUE) >= 10 ) >= 2
table(host_mrm_filt)
host_dds_filt2 <- host_dds_raw[host_mrm_filt,]

sym_dds_raw <- estimateSizeFactors(sym_dds_raw)
sym_mrm_filt <- rowSums(counts(sym_dds_raw, normalized = TRUE) >= 10 ) >= 2
table(sym_mrm_filt)
sym_dds_filt2 <- sym_dds_raw[sym_mrm_filt]
```

5. Exploratory analysis and visualization

5.1. Normalisation of gene counts
```{r}
#Option 1: normalization by trimmed mean of M values (TMM) using calcNormFactors fromn edgeR
host_norm_list <- calcNormFactors(host_dds_filt1)
host_norm_list$samples

host_norm_list <- calcNormFactors(sym_dds_filt1)
host_norm_list$samples

#Option 2: retrieve normalized counts matrix using counts function from DESeq2
host_dds_filt2 <- estimateSizeFactors(host_dds_filt2)
sizeFactors(host_dds_filt2)
host_norm_counts <- round(counts(host_dds_filt2, normalized = TRUE))
host_dds_norm <- DESeqDataSetFromMatrix(countData = host_norm_counts,
                                        colData = metadata,
                                        design = ~ RI_treatment + Temp_Col)

sym_dds_filt2 <- estimateSizeFactors(sym_dds_filt2)
sizeFactors(sym_dds_filt2)
sym_norm_counts <- round(counts(sym_dds_filt2, normalized = TRUE))
sym_dds_norm <- DESeqDataSetFromMatrix(countData = sym_norm_counts,
                                       colData = metadata,
                                       design = ~ Colony + Temperature + RI_treatment + Temperature:RI_treatment)
```

5.2. Exploration of transformation approaches
```{r}
#Performing a PCA directly with the normalized count matrices would typically result in PCAs dependent mostly on genes with the highest counts (since they show the largest absolute differences between samples).
#To avoid this, the normalized count matrices can be transformed using two approaches provided by DESeq2:
# - variance stabilizing transformation (VST)
# - the regularized-logarithm transformation (rlog)

#Normalized counts are transformed either using vst/rlog
#Setting blind to TRUE results in a transformation unbiased to sample condition information (i.e., unsupervised transformation)
host_vsd <- vst(host_dds_norm, blind = TRUE) 
host_rld <- rlog(host_dds_norm, blind = TRUE)
host_dds_norm <- estimateSizeFactors(host_dds_norm)
host_trfmd_df <- bind_rows(as_data_frame(assay(host_vsd)[, 1:2]) %>% mutate(transformation = "vst"),
                           as_data_frame(assay(host_rld)[, 1:2]) %>% mutate(transformation = "rlog"))

sym_vsd <- vst(sym_dds_norm, blind = TRUE)
sym_rld <- rlog(sym_dds_norm, blind = TRUE)
sym_dds_norm <- estimateSizeFactors(sym_dds_norm)
sym_trfmd_df <- bind_rows(as_data_frame(assay(sym_vsd)[, 1:2]) %>% mutate(transformation = "vst"),
                          as_data_frame(assay(sym_rld)[, 1:2]) %>% mutate(transformation = "rlog"))
  
#Scatter plots of transformed counts from two samples
colnames(host_trfmd_df)[1:2] <- c("x", "y")
host_lvls <- c("vst", "rlog")
host_trfmd_df$transformation <- factor(host_trfmd_df$transformation, levels=host_lvls)
ggplot(host_trfmd_df, aes(x = x, y = y)) + geom_hex(bins = 80) + coord_fixed() + facet_grid( . ~ transformation) 

colnames(sym_trfmd_df)[1:2] <- c("x", "y")
sym_lvls <- c("vst", "rlog")
sym_trfmd_df$transformation <- factor(sym_trfmd_df$transformation, levels=sym_lvls)
ggplot(sym_trfmd_df, aes(x = x, y = y)) + geom_hex(bins = 80) + coord_fixed() + facet_grid( . ~ transformation) 

#Differences between samples (deviation from y=x) will contribute to the distance calculations and the PCA plot
#Genes with low counts (bottom left-hand corner) seem to be excessively variable on the log scale
#The vst and rlog compress differences for the low count genes (which provide little information about DGE)
```


5.3.1. Exploration of similarities between samples - PCA
```{r}
#PCA to visualize similarity between samples and how the transformation method may affect this - host
host_rld_PCA <- plotPCA(host_rld, intgroup=c("Temperature", "Colony"), returnData=TRUE)
percentVar <- round(100 * attr(host_rld_PCA, "percentVar"))
host_rld_PCA_plot <- ggplot(host_rld_PCA, aes(PC1, PC2, color = Temperature, shape = Colony)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance"))

host_vsd_PCA <- plotPCA(host_vsd, intgroup=c("Temperature", "Colony"), returnData=TRUE)
percentVar <- round(100 * attr(host_vsd_PCA, "percentVar"))
host_vsd_PCA_plot <- ggplot(host_vsd_PCA, aes(PC1, PC2, color = Temperature, shape = Colony)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance"))

host_PCA <- ggarrange(host_rld_PCA_plot, host_vsd_PCA_plot,
                      labels = c("rlog", "vst"),
                      ncol = 2)
host_PCA

#PCA to visualize similarity between samples and how the transformation method may affect this - symbiont
sym_rld_PCA <- plotPCA(sym_rld, intgroup=c("RI_treatment", "Temperature"), returnData=TRUE)
percentVar <- round(100 * attr(sym_rld_PCA, "percentVar"))
sym_rld_PCA_plot <- ggplot(sym_rld_PCA, aes(PC1, PC2, color = RI_treatment, shape = Temperature)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance"))

sym_vsd_PCA <- plotPCA(sym_vsd, intgroup=c("RI_treatment", "Temperature"), returnData=TRUE)
percentVar <- round(100 * attr(sym_vsd_PCA, "percentVar"))
sym_vsd_PCA_plot <- ggplot(sym_vsd_PCA, aes(PC1, PC2, color = RI_treatment, shape = Temperature)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance"))

sym_PCA <- ggarrange(sym_rld_PCA_plot, sym_vsd_PCA_plot,
                      labels = c("rlog", "vst"),
                      ncol = 2)
sym_PCA
```

5.3.2. Exploration of similarities between samples - GLM-PCA
```{r}
#GLM-PCA works with raw count data (here using filtered gene count matrices)
#GLM-PCA of host gene counts
host_gpca <- glmpca(counts(host_dds_filt2), L = 2)
host_gpca.dat <- host_gpca$factors
host_gpca.dat$Temperature <- host_dds_filt2$Temperature
host_gpca.dat$Colony <- host_dds_filt2$Colony

ggplot(host_gpca.dat, aes(x = dim1, y = dim2, color = Temperature, shape = Colony)) +
  geom_point(size =3) + coord_fixed() + ggtitle("Host GLM-PCA")

#GLM-PCA of symbiont gene counts
sym_gpca <- glmpca(counts(sym_dds_filt2), L = 2)
sym_gpca.dat <- sym_gpca$factors
sym_gpca.dat$Temperature <- sym_dds_filt2$Temperature
sym_gpca.dat$RI_treatment <- sym_dds_filt2$RI_treatment

ggplot(sym_gpca.dat, aes(x = dim1, y = dim2, color = RI_treatment, shape = Temperature)) +
  geom_point(size =3) + coord_fixed() + ggtitle("Symbiont GLM-PCA")
```

6. DGE analysis

6.1.DGE analysis and QC
```{r}
#Run the differential expression pipeline (use filtered gene count matrices, not normalized ones as this step takes place in the process)
host_dds_dge <- DESeq(host_dds_filt2)
sym_dds_dge <- DESeq(sym_dds_filt2)

#Dispersion plots of gene-wise estimates (QC of model fitting)
plotDispEsts(host_dds_dge)
plotDispEsts(sym_dds_dge)
#Data should scatter around the fitted curve, with dispersion decreasing as mean expression levels increase
```

6.2.Host DGE analysis

6.2.1 Temperature contrast at the colony level
```{r}
#Set contrasts to be tested in pairwise comparisons
host_contrast1 <- c("Temp_Col", "Elevated_I", "Ambient_I") #The last input is the level used as baseline
host_contrast2 <- c("Temp_Col", "Elevated_J", "Ambient_J")
host_contrast3 <- c("Temp_Col", "Elevated_L", "Ambient_L")

#Build the results tables
host_res_unshrunken1 <- results(host_dds_dge, contrast = host_contrast1, alpha = 0.05)
host_res_unshrunken2 <- results(host_dds_dge, contrast = host_contrast2, alpha = 0.05)
host_res_unshrunken3 <- results(host_dds_dge, contrast = host_contrast3, alpha = 0.05)

summary(host_res_unshrunken1)
summary(host_res_unshrunken2)
summary(host_res_unshrunken3)

#Apply LFC shrinkage to the results tables (reduced noise from low gene counts)
host_res_Ash1 <- lfcShrink(host_dds_dge, contrast = host_contrast1, res= host_res_unshrunken1, type= "ashr") 
host_res_Ash2 <- lfcShrink(host_dds_dge, contrast = host_contrast2, res= host_res_unshrunken2, type= "ashr")
host_res_Ash3 <- lfcShrink(host_dds_dge, contrast = host_contrast3, res= host_res_unshrunken3, type= "ashr")
#Both apeglm and ashr are recommended over normal for LFC shrinkage (https://academic.oup.com/bioinformatics/article/35/12/2084/5159452)
#Ashr rather than apeglm is used here since it allows the use of contrast as an argument

#Pass the results from the LFC shrinked tables to dfs
host_res1 <- host_res_Ash1 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble()
host_res2 <- host_res_Ash2 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble()
host_res3 <- host_res_Ash3 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble()

mcols(host_res_Ash1, use.names = TRUE) #To extract information regarding the meaning of each column

#Set and apply thresholds to extract LFCs with meaningful biological relevance
padj.cutoff <- 0.05
lfc.cutoff <- 0.58 #Corresponds to a fold change of 1.5 in expression - appears to be a standard value used

host_signif_genes1 <- host_res1 %>% filter(padj < padj.cutoff & abs(log2FoldChange) >= lfc.cutoff)
host_signif_genes2 <- host_res2 %>% filter(padj < padj.cutoff & abs(log2FoldChange) >= lfc.cutoff)
host_signif_genes3 <- host_res3 %>% filter(padj < padj.cutoff & abs(log2FoldChange) >= lfc.cutoff)
```

6.2.2. Visualization of results - MA plots
```{r}
#MA plots displaying LFCs prior to low fold change cutoff 
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-5,5)
DESeq2::plotMA(host_res_Ash1, xlim=xlim, ylim=ylim, main="Colony I: Elevated vs. Ambient") %>% abline(h=c(-0.58,0.58), col="red", lwd=2)
DESeq2::plotMA(host_res_Ash2, xlim=xlim, ylim=ylim, main="Colony J: Elevated vs. Ambient") %>% abline(h=c(-0.58,0.58), col="red", lwd=2)
DESeq2::plotMA(host_res_Ash3, xlim=xlim, ylim=ylim, main="Colony L: Elevated vs. Ambient") %>% abline(h=c(-0.58,0.58), col="red", lwd=2)
```

6.2.3. Visualization of results - volcano plots
```{r}
#Volcano plots displaying LFCs prior to low fold change cutoff 
host_volcano1 <- EnhancedVolcano(host_res_Ash1, 
                                 lab = rownames(host_res_Ash1),
                                 x = 'log2FoldChange',
                                 y = 'pvalue',
                                 title = 'Colony I: Elevated vs. Ambient',
                                 FCcutoff = 0.58,
                                 pointSize = 1.0,
                                 labSize = 6.0)

host_volcano2 <- EnhancedVolcano(host_res_Ash2, 
                                 lab = rownames(host_res_Ash2),
                                 x = 'log2FoldChange',
                                 y = 'pvalue',
                                 title = 'Colony J: Elevated vs. Ambient',
                                 FCcutoff = 0.58,
                                 pointSize = 1.0,
                                 labSize = 6.0)

host_volcano3 <- EnhancedVolcano(host_res_Ash3, 
                                 lab = rownames(host_res_Ash3),
                                 x = 'log2FoldChange',
                                 y = 'pvalue',
                                 title = 'Colony L: Elevated vs. Ambient',
                                 FCcutoff = 0.58,
                                 pointSize = 1.0,
                                 labSize = 6.0)
host_volcano1
host_volcano2
host_volcano3
```

6.3.Symbiont DGE analysis

6.3.1 RI_treatment contrast
```{r}
#Set contrasts to be tested in pairwise comparisons
sym_contrast1 <- c("RI_treatment", "RIC", "C") #The last input is the level used as baseline
sym_contrast2 <- c("RI_treatment", "WT10", "C")
sym_contrast3 <- c("RI_treatment", "SS8", "C")
sym_contrast4 <- c("RI_treatment", "SS1", "C")

#Build the results tables
sym_res_unshrunken1 <- results(sym_dds_dge, contrast = sym_contrast1, alpha = 0.05)
sym_res_unshrunken2 <- results(sym_dds_dge, contrast = sym_contrast2, alpha = 0.05)
sym_res_unshrunken3 <- results(sym_dds_dge, contrast = sym_contrast3, alpha = 0.05)
sym_res_unshrunken4 <- results(sym_dds_dge, contrast = sym_contrast4, alpha = 0.05)

summary(sym_res_unshrunken1)
summary(sym_res_unshrunken2)
summary(sym_res_unshrunken3)
summary(sym_res_unshrunken4)

#Apply LFC shrinkage to the results tables (reduced noise from low gene counts)
sym_res_Ash1 <- lfcShrink(sym_dds_dge, contrast = sym_contrast1, res= sym_res_unshrunken1, type= "ashr") 
sym_res_Ash2 <- lfcShrink(sym_dds_dge, contrast = sym_contrast2, res= sym_res_unshrunken2, type= "ashr")
sym_res_Ash3 <- lfcShrink(sym_dds_dge, contrast = sym_contrast3, res= sym_res_unshrunken3, type= "ashr")
sym_res_Ash4 <- lfcShrink(sym_dds_dge, contrast = sym_contrast4, res= sym_res_unshrunken4, type= "ashr")

#Pass the results from the LFC shrinked tables to dfs
sym_res1 <- sym_res_Ash1 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble()
sym_res2 <- sym_res_Ash2 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble()
sym_res3 <- sym_res_Ash3 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble()
sym_res4 <- sym_res_Ash4 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble()

#Set and apply thresholds to extract LFCs with meaningful biological relevance
padj.cutoff <- 0.05
lfc.cutoff <- 0.58 #Corresponds to a fold change of 1.5 in expression - appears to be a standard value used

sym_signif_genes1 <- sym_res1 %>% filter(padj < padj.cutoff & abs(log2FoldChange) >= lfc.cutoff)
sym_signif_genes2 <- sym_res2 %>% filter(padj < padj.cutoff & abs(log2FoldChange) >= lfc.cutoff)
sym_signif_genes3 <- sym_res3 %>% filter(padj < padj.cutoff & abs(log2FoldChange) >= lfc.cutoff)
sym_signif_genes4 <- sym_res4 %>% filter(padj < padj.cutoff & abs(log2FoldChange) >= lfc.cutoff)
```

6.3.2. Visualization of results - MA plots
```{r}
#MA plots displaying LFCs prior to low fold change cutoff 
par(mfrow=c(1,4), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-5,5)
DESeq2::plotMA(sym_res_Ash1, xlim=xlim, ylim=ylim, main="RI treatment: RIC vs. C") %>% abline(h=c(-0.58,0.58), col="red", lwd=2)
DESeq2::plotMA(sym_res_Ash2, xlim=xlim, ylim=ylim, main="RI treatment: WT10 vs. C") %>% abline(h=c(-0.58,0.58), col="red", lwd=2)
DESeq2::plotMA(sym_res_Ash3, xlim=xlim, ylim=ylim, main="RI treatment: SS8 vs. C") %>% abline(h=c(-0.58,0.58), col="red", lwd=2)
DESeq2::plotMA(sym_res_Ash4, xlim=xlim, ylim=ylim, main="RI treatment: SS1 vs. C") %>% abline(h=c(-0.58,0.58), col="red", lwd=2)
```

6.3.3. Visualization of results - volcano plots
```{r}
#Volcano plots displaying LFCs prior to low fold change cutoff 
sym_volcano1 <- EnhancedVolcano(sym_res_Ash1, 
                                 lab = rownames(sym_res_Ash1),
                                 x = 'log2FoldChange',
                                 y = 'pvalue',
                                 title = 'RI treatment: RIC vs. C',
                                 FCcutoff = 0.58,
                                 pointSize = 1.0,
                                 labSize = 6.0)

sym_volcano2 <- EnhancedVolcano(sym_res_Ash2, 
                                 lab = rownames(sym_res_Ash2),
                                 x = 'log2FoldChange',
                                 y = 'pvalue',
                                 title = 'RI treatment: WT10 vs. C',
                                 FCcutoff = 0.58,
                                 pointSize = 1.0,
                                 labSize = 6.0)

sym_volcano3 <- EnhancedVolcano(sym_res_Ash3, 
                                 lab = rownames(sym_res_Ash3),
                                 x = 'log2FoldChange',
                                 y = 'pvalue',
                                 title = 'RI treatment: SS8 vs. C',
                                 FCcutoff = 0.58,
                                 pointSize = 1.0,
                                 labSize = 6.0)

sym_volcano4 <- EnhancedVolcano(sym_res_Ash4, 
                                 lab = rownames(sym_res_Ash4),
                                 x = 'log2FoldChange',
                                 y = 'pvalue',
                                 title = 'RI treatment: SS1 vs. C',
                                 FCcutoff = 0.58,
                                 pointSize = 1.0,
                                 labSize = 6.0)
sym_volcano1
sym_volcano2
sym_volcano3
sym_volcano4
```

6.3.4 RI_treatment contrast at elevated temperature
```{r}
#Change design of dds object and rerun DESeq

```