---
title: "RNAseq analysis - photosymbionts"
author: "Hugo Scharfenstein"
date: "`r Sys.Date()`"
output: html_document
---

Packages used for analysis
```{r}
#General packages
library(tidyverse)
library(reshape2)
library(vegan)
library(readxl)
library(writexl)
#DGE analysis packages
library(DESeq2)
library(edgeR)
library(glmpca)
library(apeglm)
library(ashr)
#ORA packages
library(goseq)
#Plotting packages
library(EnhancedVolcano)
library(PCAtools)
library(DEGreport)
library(RColorBrewer)
library(ggpubr)
library(ggridges)
library(plotly)
library(VennDiagram)
#Other packages
library(RColorBrewer)
library(ComplexHeatmap)
library(circlize)
library(digest)
library(cluster)
library(magick)
library(rstatix)
```

# Differential gene expression analysis - Deseq2

## Import gene count matrices generated from STAR and associated metadata
```{r}
#Import metadata, remove problematic samples
sym_metadata.df <- read_xlsx("./RNAseq dataset.xlsx", sheet = 5) %>% 
  filter(!(SampleName %in% c('Amb_RIC_J_1028_S21_'))) %>% #Remove RIC sample that contains unknown source of C. proliferum
  filter(!(SampleName %in% c('Ele_RIC_L_1290_S47_'))) %>% #Remove sample that has low read depth (<1M)
  filter(!(SampleName %in% c('Ele_SS8_L_1322_S39_'))) %>% #Remove sample that may contain mostly Durusdinium
  filter(!(SampleName %in% c('Ele_SS1_L_1334_S9_')))      #Remove outlier

#Import the gene count matrix resulting from the mapping of sequences to the C. proliferum reference genome, remove same samples as above
sym_counts.df <- read_tsv("/Users/hugo/Documents/Melbourne PhD/Experimental work/Chapter 5/Gene counts matrices/gene_counts_matrix_cprol.txt") %>%
  dplyr::select(-c(Amb_RIC_J_1028_S21_, Ele_RIC_L_1290_S47_, Ele_SS8_L_1322_S39_, Ele_SS1_L_1334_S9_))
```

## Collate exon counts to gene level
```{r}
sym_gene_counts.df <- sym_counts.df %>% 
  separate(gene_id, c("gene", "exon"), sep = "\\.exon") %>% 
  group_by(gene) %>% 
  summarise_if(is.numeric, sum)
```

## Preparation of gene count matrices for downstream analysis

### Load raw gene counts into DESeqDataSet objects 
```{r}
#Modify the metadata table, adding  sample names as row names (needed for some of the DESeq2 functions)
sym_metadata.df <- as.data.frame(sym_metadata.df)
rownames(sym_metadata.df) <- sym_metadata.df$SampleName

#Modify the sym_gene_counts.df tables, adding  gene ids as row names (needed for later)
sym_gene_counts.df <- as.data.frame(sym_gene_counts.df)
rownames(sym_gene_counts.df) <- sym_gene_counts.df$gene
sym_gene_counts.df <- sym_gene_counts.df[,-1]

#Check if row and column names match between sym_metadata.df and sym_gene_counts.df
all(colnames(sym_gene_counts.df) %in% rownames(sym_metadata.df)) 
all(colnames(sym_gene_counts.df) == rownames(sym_metadata.df))

#Load the gene count matrices from STAR into a DESeq DESeqDataSet objects
sym.dds <- DESeqDataSetFromMatrix(countData = sym_gene_counts.df,
                                  colData = sym_metadata.df,
                                  design = ~ Colony + Temp_treatment + Cprol_abund_strain + Temp_treatment:Cprol_abund_strain)

#Design matrix incorporates colony as a random factor
#Temperature, dominant photosymbiont, C. proliferum lineage and their interaction are the main factors of interest
#Dominant symbionts categories are based on whether more C1 or C21 is present in the respective samples
#For guidance on design matrices check: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7873980/
```

### Filter out lowly expressed genes (genes where normalised counts are >= 5 across at least 1 sample are kept)
```{r}
sym.dds <- estimateSizeFactors(sym.dds)
sym.dds_filt <- rowSums(counts(sym.dds, normalized = TRUE) >= 5) >= 1
table(sym.dds_filt)
sym.dds_filt2 <- sym.dds[sym.dds_filt]
```

## Exploratory analysis

### VST transformation
```{r}
sym.vsd <- vst(sym.dds_filt2, blind = F)
sym.vsd.mat <- assay(sym.vsd)
```

### Exploration of similarities between samples - PCA with PCAtools::pca
```{r}
sym.PCA_tools <- PCAtools::pca(sym.vsd.mat, metadata = sym_metadata.df, removeVar = 0.1)
sym.PCA_tools$metadata$Cprol_abundance <- factor(sym.PCA_tools$metadata$Cprol_abundance, 
                                                 levels = c("Absent", "Present", "Dominant"))
sym.PCA_tools$metadata$Cprol_strains <- factor(sym.PCA_tools$metadata$Cprol_strains, 
                                                 levels = c("SS", "WT10", "C21"))

#Screeplot to identiy optimal number of PCs
sym_horn <- parallelPCA(sym.vsd.mat)
sym_horn$n

sym_elbow <- findElbowPoint(sym.PCA_tools$variance)
sym_elbow

sym.PCA_tools.screeplot <- screeplot(sym.PCA_tools,
                                     components = getComponents(sym.PCA_tools, 1:20),
                                     vline = c(sym_horn$n, sym_elbow)) +
  geom_label(aes(x = sym_horn$n + 1, y = 50,
                 label = 'Horn\'s', vjust = -1, size = 12)) +
  geom_label(aes(x = sym_elbow + 1, y = 50,
                 label = 'Elbow method', vjust = -1, size = 12))
sym.PCA_tools.screeplot

#Eigencorplot
sym.PCA_tools.eigenplot <- eigencorplot(sym.PCA_tools,
                                        metavars = c('Cprol_strains','Cprol_abundance','Temp_treatment','Colony', 'Mapping_depth_sym'))
sym.PCA_tools.eigenplot

#Biplot - ellipses by strain
sym.PCA_tools.biplot <- biplot(sym.PCA_tools,
                               x = "PC1",
                               y = "PC2",
                               lab = NA,
                               colby = 'Cprol_abundance',
                               shape = 'Temp_treatment',
                               pointSize = 3,
                               hline = 0, hlineWidth = 0.25,
                               vline = 0, vlineWidth = 0.25,
                               xlim = c(-80, 100),
                               ylim = c(-40, 80)) +
  #geom_point(aes(alpha = sym.PCA_tools$metadata$Cprol_strains), size = 2.25, fill = 'black', colour = 'black') +
  stat_ellipse(aes(group = sym.PCA_tools$metadata$Temp_cprol_abund, fill = sym.PCA_tools$metadata$Cprol_abundance),
             geom = "polygon", alpha = 0.25, level = 0.95, type = 't') +
  stat_ellipse(aes(group = sym.PCA_tools$metadata$Temp_cprol_abund, linetype = sym.PCA_tools$metadata$Temp_treatment),
             level = 0.95, type = 't', size = 0.25, colour = 'black') +
  scale_fill_manual(values = c('#ede0d4', '#b08968', '#65451F')) +
  scale_colour_manual('C. proliferum relative abundance', values = c('#ede0d4', '#b08968', '#65451F'), 
                      labels = c('Absent', 'Under 50%', 'Over 50%')) +
  scale_linetype_manual(values = c('solid','longdash')) +
  scale_shape_manual(values = c('circle', 'square')) +
  #scale_alpha_manual('C. proliferum lineage', values = c(1, 0.5, 0), labels = c('SS', 'WT', '')) +
  theme_bw() +
  theme(panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank(),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.box = "horizontal",
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  guides(fill = F, linetype = F, 
         colour = guide_legend(title.position = "top", nrow = 1, override.aes = list(size = 3), order = 1),
         alpha = guide_legend(title.position = "top", nrow = 1, override.aes = list(size = 3), order = 2),
         shape = guide_legend(title.position = "top", override.aes = list(size = 3), order = 3)) +
  labs(shape = 'Temperature treatment')
sym.PCA_tools.biplot

Fig.3B <- sym.PCA_tools.biplot
```

### PERMANOVA
```{r}
sym.adonis <- set.seed(1234); adonis2(t(sym.vsd.mat) ~ Cprol_abundance * Temp_treatment,
                                      permutations = 10000,
                                      data = sym_metadata.df,
                                      method = "bray")

sym.adonis <- set.seed(1234); adonis2(t(sym.vsd.mat) ~ Cprol_abundance * Cprol_strains,
                                      permutations = 10000,
                                      data = sym_metadata.df,
                                      method = "bray")

sym.adonis_col <- set.seed(1234); adonis2(t(sym.vsd.mat) ~ Colony,
                                      permutations = 10000,
                                      data = sym_metadata.df,
                                      method = "bray")
```

## DGE analysis

### DGE analysis and QC
```{r}
sym.dds_unfilt <- DESeqDataSetFromMatrix(countData = sym_gene_counts.df,
                                         colData = sym_metadata.df,
                                         design = ~ Colony + Temp_cprol_abund_strain)

sym.dds_unfilt <- estimateSizeFactors(sym.dds_unfilt)
sym.dds_filt3 <- rowSums(counts(sym.dds_unfilt, normalized = TRUE) >= 5) >= 1
table(sym.dds_filt3)
sym.dds_filt4 <- sym.dds_unfilt[sym.dds_filt3]

#Run the differential expression pipeline (use filtered gene count matrices, not normalized ones as this step takes place in the process)
sym.dge <- DESeq(sym.dds_filt4)

#Dispersion plots of gene-wise estimates (QC of model fitting)
plotDispEsts(sym.dge)
#Data should scatter around the fitted curve, with dispersion decreasing as mean expression levels increase
```

### Temperature contrast for each symbiont strain & abundance combination
```{r}
sym.dge_res <- DESeq2::results(sym.dge)
metadata(sym.dge_res)$filterThreshold

sym_contrast1 <- c("Temp_cprol_abund_strain", "Elevated_Dominant_SS", "Ambient_Dominant_SS")
sym_contrast2 <- c("Temp_cprol_abund_strain", "Elevated_Present_SS", "Ambient_Present_SS")
sym_contrast3 <- c("Temp_cprol_abund_strain", "Elevated_Dominant_WT10", "Ambient_Dominant_WT10")
sym_contrast4 <- c("Temp_cprol_abund_strain", "Elevated_Present_WT10", "Ambient_Present_WT10")
sym_contrast5 <- c("Temp_cprol_abund_strain", "Elevated_Absent_C21", "Ambient_Absent_C21")

sym_res_unshrunken1 <- DESeq2::results(sym.dge, contrast = sym_contrast1, alpha = 0.05)
sym_res_unshrunken2 <- DESeq2::results(sym.dge, contrast = sym_contrast2, alpha = 0.05)
sym_res_unshrunken3 <- DESeq2::results(sym.dge, contrast = sym_contrast3, alpha = 0.05)
sym_res_unshrunken4 <- DESeq2::results(sym.dge, contrast = sym_contrast4, alpha = 0.05)
sym_res_unshrunken5 <- DESeq2::results(sym.dge, contrast = sym_contrast5, alpha = 0.05)

summary(sym_res_unshrunken1)
summary(sym_res_unshrunken2)
summary(sym_res_unshrunken3)
summary(sym_res_unshrunken4)
summary(sym_res_unshrunken5)

sym_res_Ash1 <- lfcShrink(sym.dge, contrast = sym_contrast1, res= sym_res_unshrunken1, type= "ashr") 
sym_res_Ash2 <- lfcShrink(sym.dge, contrast = sym_contrast2, res= sym_res_unshrunken2, type= "ashr") 
sym_res_Ash3 <- lfcShrink(sym.dge, contrast = sym_contrast3, res= sym_res_unshrunken3, type= "ashr") 
sym_res_Ash4 <- lfcShrink(sym.dge, contrast = sym_contrast4, res= sym_res_unshrunken4, type= "ashr") 
sym_res_Ash5 <- lfcShrink(sym.dge, contrast = sym_contrast5, res= sym_res_unshrunken5, type= "ashr") 

sym_shrink_res1 <- sym_res_Ash1 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble() %>%
  mutate(contrast = 'Dominant_SS')
sym_shrink_res2 <- sym_res_Ash2 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble() %>%
  mutate(contrast = 'Present_SS')
sym_shrink_res3 <- sym_res_Ash3 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble() %>%
  mutate(contrast = 'Dominant_WT10')
sym_shrink_res4 <- sym_res_Ash4 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble() %>%
  mutate(contrast = 'Present_WT10')
sym_shrink_res5 <- sym_res_Ash5 %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble() %>%
  mutate(contrast = 'Absent_C21')

padj.cutoff <- 0.05
lfc.cutoff <- 0.58

sym_signif_genes1 <- sym_shrink_res1 %>% dplyr::filter(padj < padj.cutoff) #& abs(log2FoldChange) >= lfc.cutoff)
sym_signif_genes2 <- sym_shrink_res2 %>% dplyr::filter(padj < padj.cutoff) #& abs(log2FoldChange) >= lfc.cutoff)
sym_signif_genes3 <- sym_shrink_res3 %>% dplyr::filter(padj < padj.cutoff) #& abs(log2FoldChange) >= lfc.cutoff)
sym_signif_genes4 <- sym_shrink_res4 %>% dplyr::filter(padj < padj.cutoff) #& abs(log2FoldChange) >= lfc.cutoff)
sym_signif_genes5 <- sym_shrink_res5 %>% dplyr::filter(padj < padj.cutoff) #& abs(log2FoldChange) >= lfc.cutoff)

sym.DEGs <- rbind(sym_shrink_res1, sym_shrink_res2) %>% 
  rbind(sym_shrink_res3) %>%
  rbind(sym_shrink_res4) %>%
  rbind(sym_shrink_res5)
```

### Visualisation of expression levels of DEGs
```{r}
sym.DEGs_SS <- sym.DEGs %>%
  filter(contrast == 'Dominant_SS') %>%
  filter(padj < 0.05) %>% #& abs(log2FoldChange) >= lfc.cutoff)
  column_to_rownames(var = 'gene')

sym.DEGs_WT10 <- sym.DEGs %>% 
  filter(contrast == 'Dominant_WT10') %>%
  filter(padj < 0.05) %>% #& abs(log2FoldChange) >= lfc.cutoff)
  column_to_rownames(var = 'gene')

sym.DEGs_C21 <- sym.DEGs %>%
  filter(contrast == 'Absent_C21') %>%
  filter(padj < 0.05) %>% #& abs(log2FoldChange) >= lfc.cutoff)
  column_to_rownames(var = 'gene')

#Define different groups that will be coloured and shaped differently
group1.col <- c('Dominant_SS')
group2.col <- c('Dominant_WT10')
group3.col <- c('Absent_C21')

#Create custom key-value pairs for different GO terms and photosymbiont groups
keyvals.col1 <- ifelse(sym.DEGs_SS$contrast %in% group1.col, '#DB8941',
                       'black')
keyvals.col2 <- ifelse(sym.DEGs_WT10$contrast %in% group2.col, '#9f6671',
                       'black')
keyvals.col3 <- ifelse(sym.DEGs_C21$contrast %in% group3.col, '#0e3b5c',
                       'black')

keyvals.col1[is.na(keyvals.col1)] <- 'black'
names(keyvals.col1)[keyvals.col1 == '#DB8941'] <- 'SS'
keyvals.col2[is.na(keyvals.col2)] <- 'black'
names(keyvals.col2)[keyvals.col2 == '#9f6671'] <- 'WT10'
keyvals.col3[is.na(keyvals.col3)] <- 'black'
names(keyvals.col3)[keyvals.col3 == '#0e3b5c'] <- 'C21'

#Identify DEGs that are above lgFC of 6 or -6
sym.lfc_cutoff <- 6
up_SS <- rownames(subset(sym.DEGs_SS, log2FoldChange >= sym.lfc_cutoff))
down_SS <- rownames(subset(sym.DEGs_SS, log2FoldChange <= sym.lfc_cutoff * -1))
up_WT10 <- rownames(subset(sym.DEGs_WT10, log2FoldChange >= sym.lfc_cutoff))
down_WT10 <- rownames(subset(sym.DEGs_WT10, log2FoldChange <= sym.lfc_cutoff * -1))
up_C21 <- rownames(subset(sym.DEGs_C21, log2FoldChange >= sym.lfc_cutoff))
down_C21 <- rownames(subset(sym.DEGs_C21, log2FoldChange <= sym.lfc_cutoff * -1))

#Impute l2FCs in the object to be max sym.lfc_cutoff or min -sym.lfc_cutoff 
sym.DEGs_SS$log2FoldChange <- ifelse(sym.DEGs_SS$log2FoldChange >= sym.lfc_cutoff, sym.lfc_cutoff,
                                     ifelse(sym.DEGs_SS$log2FoldChange <= sym.lfc_cutoff * -1, sym.lfc_cutoff * -1,
                                            sym.DEGs_SS$log2FoldChange))
max(sym.DEGs_SS$log2FoldChange, na.rm = TRUE)
min(sym.DEGs_SS$log2FoldChange, na.rm = TRUE)

sym.DEGs_WT10$log2FoldChange <- ifelse(sym.DEGs_WT10$log2FoldChange >= sym.lfc_cutoff, sym.lfc_cutoff,
                                     ifelse(sym.DEGs_WT10$log2FoldChange <= sym.lfc_cutoff * -1, sym.lfc_cutoff * -1,
                                            sym.DEGs_WT10$log2FoldChange))
max(sym.DEGs_WT10$log2FoldChange, na.rm = TRUE)
min(sym.DEGs_WT10$log2FoldChange, na.rm = TRUE)

sym.DEGs_C21$log2FoldChange <- ifelse(sym.DEGs_C21$log2FoldChange >= sym.lfc_cutoff, sym.lfc_cutoff,
                                     ifelse(sym.DEGs_C21$log2FoldChange <= sym.lfc_cutoff * -1, sym.lfc_cutoff * -1,
                                            sym.DEGs_C21$log2FoldChange))
max(sym.DEGs_C21$log2FoldChange, na.rm = TRUE)
min(sym.DEGs_C21$log2FoldChange, na.rm = TRUE)

#Custom shapes for the points
customshape_SS <- rep(19, nrow(sym.DEGs_SS))
names(customshape_SS) <- rep('Group1', nrow(sym.DEGs_SS))
customshape_SS[which(rownames(sym.DEGs_SS) %in% up_SS)] <- -9658
names(customshape_SS)[which(rownames(sym.DEGs_SS) %in% up_SS)] <- 'Group2'
customshape_SS[which(rownames(sym.DEGs_SS) %in% down_SS)] <- -9668
names(customshape_SS)[which(rownames(sym.DEGs_SS) %in% down_SS)] <- 'Group3'

customshape_WT10 <- rep(19, nrow(sym.DEGs_WT10))
names(customshape_WT10) <- rep('Group1', nrow(sym.DEGs_WT10))
customshape_WT10[which(rownames(sym.DEGs_WT10) %in% up_WT10)] <- -9658
names(customshape_WT10)[which(rownames(sym.DEGs_WT10) %in% up_WT10)] <- 'Group2'
customshape_WT10[which(rownames(sym.DEGs_WT10) %in% down_WT10)] <- -9668
names(customshape_WT10)[which(rownames(sym.DEGs_WT10) %in% down_WT10)] <- 'Group3'

customshape_C21 <- rep(19, nrow(sym.DEGs_C21))
names(customshape_C21) <- rep('Group1', nrow(sym.DEGs_C21))
customshape_C21[which(rownames(sym.DEGs_C21) %in% up_C21)] <- -9658
names(customshape_C21)[which(rownames(sym.DEGs_C21) %in% up_C21)] <- 'Group2'
customshape_C21[which(rownames(sym.DEGs_C21) %in% down_C21)] <- -9668
names(customshape_C21)[which(rownames(sym.DEGs_C21) %in% down_C21)] <- 'Group3'

#Custom sizes for the points
customsize_SS <- rep(1, nrow(sym.DEGs_SS))
customsize_SS [which(rownames(sym.DEGs_SS) %in% up_SS)] <- 2
customsize_SS [which(rownames(sym.DEGs_SS) %in% down_SS)] <- 2

customsize_WT10 <- rep(1, nrow(sym.DEGs_WT10))
customsize_WT10 [which(rownames(sym.DEGs_WT10) %in% up_WT10)] <- 2
customsize_WT10 [which(rownames(sym.DEGs_WT10) %in% down_WT10)] <- 2

customsize_C21 <- rep(1, nrow(sym.DEGs_C21))
customsize_C21 [which(rownames(sym.DEGs_C21) %in% up_C21)] <- 2
customsize_C21 [which(rownames(sym.DEGs_C21) %in% down_C21)] <- 2

#Volcano plots
sym.DEGS.volc_plot_SS <- EnhancedVolcano(sym.DEGs_SS, lab = NA,
                                         x = 'log2FoldChange', y = 'padj',
                                         xlim = c(-6, 6), #ylim = c(0, -log10(10e-52)),
                                         title = '', subtitle = 'C. proliferum (SS)',
                                         caption = paste0(nrow(sym.DEGs_SS), " DEGs"),
                                         legendPosition = "none", legendLabSize = 8, legendIconSize = 2,
                                         pCutoff = 0.05, FCcutoff = 0.58,
                                         cutoffLineWidth = 0.25, cutoffLineCol = "black",
                                         cutoffLineType = "dashed",
                                         colCustom = keyvals.col1,
                                         shapeCustom = customshape_SS,
                                         pointSize = customsize_SS,
                                         gridlines.minor = F, gridlines.major = F,
                                         labSize = 3.0, axisLabSize = 14, colAlpha = 1) + 
  theme(plot.margin=unit(c(-1,-0.25,-0.95,0), "cm")) +
  guides(colour = F)

sym.DEGS.volc_plot_WT10 <- EnhancedVolcano(sym.DEGs_WT10, lab = NA,
                                           x = 'log2FoldChange', y = 'padj',
                                           xlim = c(-6, 6), #ylim = c(0, -log10(10e-52)),
                                           title = '', subtitle = 'C. proliferum (WT)', 
                                           caption = paste0(nrow(sym.DEGs_WT10), " DEGs"),
                                           legendPosition = "none", legendLabSize = 8, legendIconSize = 2,
                                           pCutoff = 0.05, FCcutoff = 0.58,
                                           cutoffLineWidth = 0.25, cutoffLineCol = "black",
                                           cutoffLineType = "dashed",
                                           colCustom = keyvals.col2,
                                           shapeCustom = customshape_WT10,
                                           pointSize = customsize_WT10,
                                           gridlines.minor = F, gridlines.major = F,
                                           labSize = 3.0, axisLabSize = 14, colAlpha = 1) + 
  theme(plot.margin=unit(c(-1,-0.25,-0.95,0), "cm")) +
  guides(colour = F)

sym.DEGS.volc_plot_C21 <- EnhancedVolcano(sym.DEGs_C21, lab = NA,
                                          x = 'log2FoldChange', y = 'padj',
                                          xlim = c(-6, 6), #ylim = c(0, -log10(10e-52)),
                                          title = '', subtitle = 'C21', 
                                          caption = paste0(nrow(sym.DEGs_C21), " DEGs"),
                                          legendPosition = "none", legendLabSize = 8, legendIconSize = 2,
                                          pCutoff = 0.05, FCcutoff = 0.58,
                                          cutoffLineWidth = 0.25, cutoffLineCol = "black", 
                                          cutoffLineType = "dashed",
                                          colCustom = keyvals.col3,
                                          shapeCustom = customshape_C21,
                                          pointSize = customsize_C21,
                                          gridlines.minor = F, gridlines.major = F,
                                          labSize = 3.0, axisLabSize = 14, colAlpha = 1) + 
  theme(plot.margin=unit(c(-1,-0.25,-0.95,0), "cm")) +
  guides(colour = F)

sym.DEG.volc_plots <- ggarrange(sym.DEGS.volc_plot_SS + rremove('ylab') + rremove('xlab'),
                                sym.DEGS.volc_plot_WT10 + rremove('ylab') + rremove('xlab'),
                                sym.DEGS.volc_plot_C21 + rremove('ylab') + rremove('xlab'),
                                #common.legend = T,
                                #legend = "top",
                                heights = c(1, 1, 1),
                                align = 'hv',
                                ncol = 3)
sym.DEG.volc_plots

sym.DEG.volc_plots_annot <- annotate_figure(sym.DEG.volc_plots,
                                            bottom = text_grob(bquote(bold(~Log[2]~"Fold Change")),
                                                               color = "black", face = "bold", size = 14),
                                            left = text_grob(bquote(bold(~-Log[10] ~ italic(P))),
                                                             color= "black",face = "bold",size = 14,rot = 90))
sym.DEG.volc_plots_annot

Fig.5B <- sym.DEG.volc_plots_annot
```

### Visualisation of expression levels of DEGs - barplot
```{r}
#Calculate mean expression levels for down- and up-regulated DEGs
sym.DEGs_down <- sym.DEGs %>%
  group_by(contrast) %>%
  filter(padj < 0.05) %>%
  filter(!(contrast %in% c('Present_SS', 'Present_WT10'))) %>%
  filter(log2FoldChange < 0) %>%
  mutate(mean_l2fc = mean(log2FoldChange)) %>%
  mutate(se_l2fc = sd(log2FoldChange)/sqrt(length((log2FoldChange)))) #%>%
  #mutate(direction = '-') %>%
  #select(contrast, mean_l2fc, se_l2fc, direction) %>%
  #distinct()

sym.DEGs_up <- sym.DEGs %>%
  group_by(contrast) %>%
  filter(padj < 0.05) %>%
  filter(!(contrast %in% c('Present_SS', 'Present_WT10'))) %>%
  filter(log2FoldChange > 0) %>%
  mutate(mean_l2fc = mean(log2FoldChange)) %>%
  mutate(se_l2fc = sd(log2FoldChange)/sqrt(length((log2FoldChange)))) #%>%
  #mutate(direction = '+') %>%
  #select(contrast, mean_l2fc, se_l2fc, direction) %>%
  #distinct()

#Check for significant differences between photosymbiont communities (using Tukey’s HSD)
tukey_hsd.lfc_down <- aov(log2FoldChange ~ contrast, data = sym.DEGs_down) %>% tukey_hsd()
tukey_hsd.lfc_up <- aov(log2FoldChange ~ contrast, data = sym.DEGs_up) %>% tukey_hsd()

#Plot barplot with mean expression levels
sym.DEGs_barplot <- rbind(sym.DEGs_down, sym.DEGs_up) %>%
  mutate(contrast = factor(contrast, levels = c("Dominant_SS", "Dominant_WT10", "Absent_C21"))) %>%
  ggplot() +
  stat_pvalue_manual(tukey_hsd.lfc_up, label = "p.adj.signif",  tip.length = 0, coord.flip = TRUE,
                     y.position = c(0.9, 1.05, 0.975)) +
  stat_pvalue_manual(tukey_hsd.lfc_down, label = "p.adj.signif",  tip.length = 0, coord.flip = TRUE,
                     vjust = 2,
                     y.position = c(-0.8, -0.95, -0.875)) +
  geom_bar(aes(x = contrast, y = mean_l2fc, fill = contrast),
               stat = 'identity', position = position_dodge(), width = 0.5) +
  geom_linerange(aes(x = contrast, ymin = mean_l2fc - se_l2fc, ymax = mean_l2fc + se_l2fc),
                 stat = 'identity', position = position_dodge(0.5), size = 3) +
  geom_hline(aes(yintercept = 0), linetype = 'dashed', size = 0.5) +
  scale_fill_manual(values = c('#DB8941', '#9f6671', '#0e3b5c')) +
  scale_y_continuous(bquote(bold(Mean~Log[2]~"Fold Change")), limits = c(-1,1.1)) +
  scale_x_discrete('Lineage', expand = c(0, 0), 
                   labels = c('C21', 'SS', 'WT')) +
  coord_flip() +
  theme_bw() +
  theme(panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank(),
        panel.border = element_blank(),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  guides(fill = F)
sym.DEGs_barplot

Fig.5C <- sym.DEGs_barplot
```

### Visualisation of unique DEGs
```{r}
sym.DEGs_up_down <- sym.DEGs %>%
  group_by(contrast) %>%
  filter(padj < 0.05) %>%
  dplyr::select(contrast, gene, log2FoldChange, padj) %>%
  mutate(direction = case_when(log2FoldChange < 0 ~ '-',
                               log2FoldChange > 0 ~ '+')) %>%
  dplyr::select(contrast, direction) %>%
  group_by(contrast, direction) %>%
  dplyr::count(contrast, direction)

sym.venn <- venn.diagram(x = list(sym_signif_genes1$gene, sym_signif_genes3$gene, sym_signif_genes5$gene),
  category.names = c("C. proliferum (SS)", "C. proliferum (WT)",  "C21"),
  filename = NULL, #Add filename here to save venn diagram
  disable.logging = TRUE,
  lwd = 2,
  lty = 1,
  col = c('#DB8941',  '#9f6671',  '#0e3b5c'),
  fill = c(alpha('#DB8941',0.5), alpha('#9f6671',0.5), alpha('#0e3b5c',0.5)),
  cex = 1.25,
  fontface = "bold",
  cat.col = c('#DB8941',  '#9f6671', '#0e3b5c'),
  cat.cex = 1.25,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-25, 15, 140),
  cat.dist = c(0.06, 0.06, 0.1),
  rotation = 1)
grid.newpage()
grid.draw(sym.venn)

Fig.5D <- sym.venn
```

### Visualisation of gene expression level for symbiont DEGs

### Subset gene count matrix with DEGs 
```{r}
#Filter for significant DEGs
sym.DEGs_signif <- sym.DEGs %>%
  filter(padj < 0.05) %>% #& abs(log2FoldChange) >= lfc.cutoff) %>% 
  dplyr::select(gene) %>% 
  distinct() %>% 
  pull()

#Keep only genes that are differentially expressed
sym_gene_counts.df.DEGs <- sym_gene_counts.df[sym.DEGs_signif,]
```

### VST transform gene counts (includes count normalisation) and scale to obtain gene z-scores
```{r}
sym.dds.DEGs <- DESeqDataSetFromMatrix(countData = sym_gene_counts.df.DEGs,
                                       colData = sym_metadata.df,
                                       design = ~ Colony + Temp_treatment + Cprol_abund_strain + Temp_treatment:Cprol_abund_strain)

sym.vsd_DEGs <- vst(sym.dds.DEGs, blind=FALSE)
sym.vsd_DEGs.mat <- assay(sym.vsd_DEGs) 
sym.vsd_DEGs_scaled <- t(scale(t(sym.vsd_DEGs.mat))) 
sym.vsd_DEGs_scaled2 <- sym.vsd_DEGs_scaled %>%
  melt() %>%
  rename(gene = Var1, SampleName = Var2, count_norm = value) %>%
  left_join(., dplyr::select(sym_metadata.df, c(SampleName, Temp_treatment, Cprol_abundance, Cprol_strains)))
sym.vsd_DEGs.mat2 <- sym.vsd_DEGs.mat %>%
  melt() %>%
  rename(gene = Var1, SampleName = Var2, count_norm = value) %>%
  left_join(., dplyr::select(sym_metadata.df, c(SampleName, Temp_treatment, Cprol_abundance, Cprol_strains)))
```

### Density plot of gene z-scores under each temperature treatment
```{r}
sym.vsd_DEGs_scaled2$Cprol_strains <- factor(sym.vsd_DEGs_scaled2$Cprol_strains, levels = c("SS", "WT10", "C21"))
sym.vsd_DEGs_scaled2$Cprol_abundance <- factor(sym.vsd_DEGs_scaled2$Cprol_abundance, 
                                               levels = c("Dominant", "Present", "Absent"))

sym.DEG_expression.ridge_plot <- sym.vsd_DEGs_scaled2 %>%
  filter(!(Cprol_abundance %in% c('Present'))) %>%
  ggplot(aes(x = count_norm, y = interaction(Cprol_strains, Cprol_abundance), fill = Temp_treatment)) +
  geom_vline(aes(xintercept = 0), linetype = 'dashed', size = 0.25) +
  geom_density_ridges(alpha = 0.5, size = 0.25) +
  scale_fill_manual('Temperature treatment', values = c('#2171B5', '#CB181D')) +
  scale_x_continuous('Gene Z-score', expand = c(0, 0), limits = c(-4, 4), breaks = -4:4*2) +
  scale_y_discrete('C. proliferum abundance (lineage)', expand = c(0, 0), 
                   labels = c('Over 50% (SS)', 'Over 50% (WT)', 'Absent (C21)')) +
  theme_bw() +
  theme(panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank(),
        panel.border = element_blank(),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black"))
sym.DEG_expression.ridge_plot

sym.DEG_expression.barplot <- sym.vsd_DEGs_scaled2 %>%
  filter(!(Cprol_abundance %in% c('Present'))) %>%
  group_by(Temp_treatment, Cprol_abundance, Cprol_strains) %>%
  mutate(abs_zscore = abs(count_norm)) %>%
  mutate(mean_zscore = mean(abs_zscore)) %>%
  mutate(se_zscore = sd(abs_zscore)/sqrt(length((abs_zscore)))) %>%
  ungroup() %>%
  dplyr::select(-c(gene, count_norm, abs_zscore)) %>%
  distinct() %>%
  ggplot(aes(x = interaction(Cprol_strains, Cprol_abundance), fill = Temp_treatment)) +
  geom_bar(aes(y = mean_zscore), stat = 'identity', position = position_dodge(), width = 0.75) +
  geom_linerange(aes(ymin = mean_zscore - se_zscore, ymax = mean_zscore + se_zscore), stat = 'identity',
                 position = position_dodge(0.75), size = 3) +
  scale_fill_manual('Temperature treatment', values = c('#2171B5', '#CB181D')) +
  scale_y_continuous('Mean absolute gene Z-score', expand = c(0,0)) +
  scale_x_discrete('C. proliferum abundance (lineage)', expand = c(0, 0), 
                   labels = c('Over 50% (SS)', 'Over 50% (WT)', 'Absent (C21)')) +
  coord_flip() +
  theme_bw() +
  theme(panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank(),
        panel.border = element_blank(),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black"))
sym.DEG_expression.barplot
```

### Heatmap of gene z-scores
```{r}
myCol <- colorRampPalette(c('#005AB5', 'white', '#DC3220'))(100)
myBreaks <- seq(-3, 3, length.out = 100)

sym_ann <- data.frame(
  Temperature = sym_metadata.df$Temp_treatment,
  Cprol_abundance = sym_metadata.df$Cprol_abundance,
  Cprol_strain = sym_metadata.df$Cprol_strains,
  stringsAsFactors = FALSE)

sym_colours <- list(
  Temperature = c('Ambient' = '#2171B5', 'Elevated' = '#CB181D'),
  Cprol_abundance = c('Dominant' = '#65451F', 'Present' = '#b08968', 'Absent' = '#ede0d4'),
  Cprol_strain = c('SS' = '#DB8941', 'WT10' = '#9f6671', 'C21' = 'white'))

sym_colAnn <- HeatmapAnnotation(
  df = sym_ann,
  which = 'col', #Col (samples) or row (gene) annotation
  na_col = 'white', #Default colour for any NA values in the annotation data-frame, 'ann'
  col = sym_colours,
  annotation_height = 0.6,
  annotation_width = unit(1, 'cm'),
  gap = unit(1, 'mm'),
  show_annotation_name = FALSE,
  annotation_legend_param = list(
    Temperature = list(
      nrow = 2, #Number of rows across which the legend will be arranged
      title = 'Temperature',
      title_position = 'topleft',
      legend_direction = 'vertical',
      title_gp = gpar(fontsize = 12, fontface = 'bold'),
      labels_gp = gpar(fontsize = 12, fontface = 'plain')),
    Cprol_abundance = list(
      nrow = 3,
      title = 'C. proliferum abundance',
      title_position = 'topleft',
      legend_direction = 'vertical',
      title_gp = gpar(fontsize = 12, fontface = 'bold'),
      labels_gp = gpar(fontsize = 12, fontface = 'plain'),
      at = c("Dominant", "Present", "Absent"),
      labels = c("Over 50%", "Under 50%", "Absent")),
    Cprol_strain = list(
      nrow = 3,
      title = 'C. proliferum lineage',
      title_position = 'topleft',
      legend_direction = 'vertical',
      title_gp = gpar(fontsize = 12, fontface = 'bold'),
      labels_gp = gpar(fontsize = 12, fontface = 'plain'),
      at = c("SS", "WT10", "C21"),
      labels = c("SS", "WT", ""))))
  
sym_boxplotCol <- HeatmapAnnotation(
  show_annotation_name = FALSE,
  boxplot = anno_boxplot(
      sym.vsd_DEGs_scaled,
      border = FALSE,
      gp = gpar(fill = '#CCCCCC'),
      pch = '.',
      size = unit(2, 'mm'),
      axis = TRUE,
      axis_param = list(
        gp = gpar(fontsize = 12),
        side = 'left')),
      annotation_width = unit(c(2.0), 'cm'),
      which = 'col')

pamClusters <- cluster::pam(sym.vsd_DEGs_scaled, k = 3) #Pre-select k = 3 centers
pamClusters$clustering <- paste0('Cluster ', pamClusters$clustering)

pamClusters$clustering <- factor(pamClusters$clustering,
                                 levels = c('Cluster 1', 'Cluster 2', 'Cluster 3'))

sym.hmap <- Heatmap(sym.vsd_DEGs_scaled,
                    split = pamClusters$clustering,
                    cluster_row_slices = FALSE,
                    name = 'Gene Z-score',
                    col = colorRamp2(myBreaks, myCol),
                    heatmap_legend_param = list(
                      color_bar = 'continuous',
                      legend_direction = 'vertical',
                      legend_width = unit(4, 'cm'),
                      legend_height = unit(4.0, 'cm'),
                      title_position = 'leftcenter-rot',
                      title_gp=gpar(fontsize = 12, fontface = 'bold'),
                      labels_gp=gpar(fontsize = 12, fontface = 'plain')),
                    cluster_rows = TRUE,
                    show_row_dend = TRUE,
                    row_title = 'Differentially expressed genes (DEGs)',
                    row_title_side = 'left',
                    row_title_gp = gpar(fontsize = 14,  fontface = 'bold'),
                    row_title_rot = 90,
                    show_row_names = FALSE,
                    row_names_gp = gpar(fontsize = 12, fontface = 'plain'),
                    row_names_side = 'left',
                    row_dend_width = unit(15,'mm'),
                    cluster_columns = TRUE,
                    show_column_dend = TRUE,
                    column_title = '',
                    column_title_side = 'bottom',
                    column_title_gp = gpar(fontsize = 14, fontface = 'bold'),
                    column_title_rot = 0,
                    show_column_names = FALSE,
                    column_names_gp = gpar(fontsize = 12, fontface = 'plain'),
                    column_names_max_height = unit(10, 'cm'),
                    column_dend_height = unit(15,'mm'),
                    clustering_distance_columns = function(x) as.dist(1 - cor(t(x))),
                    clustering_method_columns = 'ward.D2',
                    clustering_distance_rows = function(x) as.dist(1 - cor(t(x))),
                    clustering_method_rows = 'ward.D2',
                    top_annotation = sym_colAnn,
                    bottom_annotation = sym_boxplotCol,
                    width = unit(15, "cm"))

sym.DEGs_hmap = grid.grabExpr(draw(sym.hmap,
                                   heatmap_legend_side = 'left',
                                   annotation_legend_side = 'right',
                                   row_sub_title_side = 'left',
                                   align_heatmap_legend = 'global_center',
                                   align_annotation_legend = 'global_center',
                                   column_title = "",
                                   column_title_gp = gpar(fontsize = 14, fontface = 'bold')))
sym.DEGs_hmap
is.grob(sym.DEGs_hmap)
```

## Assemble plots from DGE analysis for export
```{r}
sym_DEGs.panel1 <- cowplot::plot_grid(Fig.5B,
                                      Fig.5C,
                                      Fig.5D,
                                      labels = c("B", "C", "D"),
                                      ncol = 1, align = "v",
                                      axis = "b",
                                      rel_heights = c(1, 0.5, 0.6),
                                      rel_widths = c(1, 0.75, 0.75))
sym_DEGs.panel1

sym_DEGs.panel2 <- cowplot::plot_grid(sym.DEGs_hmap,
                                      sym_DEGs.panel1,
                                      labels = c("A", ""),
                                      ncol = 2, align = "v",
                                      axis = "b",
                                      rel_heights = c(1, 1),
                                      rel_widths = c(1, 0.5))
sym_DEGs.panel2
```

### Save DGE analysis plots of interest
```{r}
ggsave(plot = sym_DEGs.panel2, "./Fig. 5.pdf", 
       path = "./",
       width = 16.5,
       height = 10,
       units = 'in',
       dpi = 300)
```

## Over-representation analysis

### Import exon lengths and gene ontology (GO) terms for C. proliferum
```{r}
#Import exon lengths of C. proliferum
cprol_transcript.length <- read_xlsx("./RNAseq dataset.xlsx", sheet = 7) %>%
  group_by(exon_id) %>%
  mutate(transcript_length = sum(length)) %>%
  dplyr::select(exon_id, transcript_length) %>%
  rename('gene' = 1, 'length' = 2) %>%
  distinct()

#Import gene counts from earlier
sym_gene_counts.df2 <-sym_gene_counts.df %>%
  rowMeans() %>%
  as.data.frame() %>%
  rownames_to_column(var = 'gene') %>%
  rename('length' = 2)

#Import annotated genes with GO terms
cprol_gene.annot <- read_xlsx("./RNAseq dataset.xlsx", sheet = 9) %>%
  rename('gene' = 1, 'protein' = 2, 'organism' = 3, 'GO_ID' = 4, 'GO' = 5)
```

### Prepare lists of background genes and DEGs - upregulated and downregulated DEGs analysed seperately
```{r}
#Run this analysis once with upregulated DEGs then repeat with downregulated DEGs
#List of background genes resulting from the DGE analysis (this means genes with low counts are filtered out)
sym_ALL.DGE_assayed <- sym.DEGs %>%
  filter(log2FoldChange > 0) %>%
  #filter(log2FoldChange < 0) %>%
  dplyr::select(gene) %>%
  distinct() %>%
  mutate_all(trimws)
sym_ALL.DGE_assayed_list <- sym_ALL.DGE_assayed[['gene']]

#List of DEGs for each contrast (filter for significant genes i.e., padj < 0.05)
sym_DEGs_assayed.1 <- sym.DEGs %>%
  filter(log2FoldChange > 0) %>%
  #filter(log2FoldChange < 0) %>%
  filter(contrast == 'Dominant_SS') %>%
  filter(padj < 0.05) %>%
  dplyr::select(gene) %>%
  distinct() %>%
  mutate_all(trimws)
sym_DEGs_assayed_list.1 <- sym_DEGs_assayed.1[['gene']]

sym_DEGs_assayed.2 <- sym.DEGs %>%
  filter(log2FoldChange > 0) %>%
  #filter(log2FoldChange < 0) %>%
  filter(contrast == 'Dominant_WT10') %>%
  filter(padj < 0.05) %>%
  dplyr::select(gene) %>%
  distinct() %>%
  mutate_all(trimws)
sym_DEGs_assayed_list.2 <- sym_DEGs_assayed.2[['gene']]

sym_DEGs_assayed.3 <- sym.DEGs %>%
  filter(log2FoldChange > 0) %>%
  #filter(log2FoldChange < 0) %>%
  filter(contrast == 'Absent_C21') %>%
  filter(padj < 0.05) %>%
  dplyr::select(gene) %>%
  distinct() %>%
  mutate_all(trimws)
sym_DEGs_assayed_list.3 <- sym_DEGs_assayed.3[['gene']]
```

### Calculate probability weighting function using goseq
```{r}
#Construct a new vector that adds a 0 next to every gene that is not in the DEG list and a 1 next to every gene that is in the DEG list
sym_gene.vector.1 = as.integer(sym_ALL.DGE_assayed_list %in% sym_DEGs_assayed_list.1)
names(sym_gene.vector.1) = sym_ALL.DGE_assayed_list

sym_gene.vector.2 = as.integer(sym_ALL.DGE_assayed_list %in% sym_DEGs_assayed_list.2)
names(sym_gene.vector.2) = sym_ALL.DGE_assayed_list

sym_gene.vector.3 = as.integer(sym_ALL.DGE_assayed_list %in% sym_DEGs_assayed_list.3)
names(sym_gene.vector.3) = sym_ALL.DGE_assayed_list

#Ensure that the list of transcript lengths matches sym_gene.vector
cprol_transcript.length.1 <- cprol_transcript.length[cprol_transcript.length$gene %in% sym_ALL.DGE_assayed$gene, ]
gene_length_cprol_list.1 <- deframe(cprol_transcript.length.1)

cprol_transcript.length.2 <- cprol_transcript.length[cprol_transcript.length$gene %in% sym_ALL.DGE_assayed$gene, ]
gene_length_cprol_list.2 <- deframe(cprol_transcript.length.2)

cprol_transcript.length.3 <- cprol_transcript.length[cprol_transcript.length$gene %in% sym_ALL.DGE_assayed$gene, ]
gene_length_cprol_list.3 <- deframe(cprol_transcript.length.3)

#Ensure order of vectors are the same and prepare vectors for pwf() function 
sym_gene.vector.1 <- sym_gene.vector.1[order(names(sym_gene.vector.1))]
gene_length_cprol_list.1 <- gene_length_cprol_list.1[order(names(gene_length_cprol_list.1))]
gene_length_cprol_list.1 <- unname(gene_length_cprol_list.1)
gene_length_cprol_list.1 <- as.integer(gene_length_cprol_list.1)

sym_gene.vector.2 <- sym_gene.vector.2[order(names(sym_gene.vector.2))]
gene_length_cprol_list.2 <- gene_length_cprol_list.2[order(names(gene_length_cprol_list.2))]
gene_length_cprol_list.2 <- unname(gene_length_cprol_list.2)
gene_length_cprol_list.2 <- as.integer(gene_length_cprol_list.2)

sym_gene.vector.3 <- sym_gene.vector.3[order(names(sym_gene.vector.3))]
gene_length_cprol_list.3 <- gene_length_cprol_list.3[order(names(gene_length_cprol_list.3))]
gene_length_cprol_list.3 <- unname(gene_length_cprol_list.3)
gene_length_cprol_list.3 <- as.integer(gene_length_cprol_list.3)

#Weigh the gene vector by lengths of the transcripts
sym_pwf.1 = nullp(sym_gene.vector.1, bias.data = gene_length_cprol_list.1)
sym_pwf.2 = nullp(sym_gene.vector.2, bias.data = gene_length_cprol_list.2)
sym_pwf.3 = nullp(sym_gene.vector.3, bias.data = gene_length_cprol_list.3)
```

### Get GO terms and carry out over-representation analysis
```{r}
#Prepare list of GO terms for each gene
sym_GO.df <- cprol_gene.annot %>% dplyr::select(gene, GO_ID) %>% separate_rows(GO_ID, sep = '; ')
sym_GO.list <- split(sym_GO.df$GO_ID, sym_GO.df$gene)

#Carry out go GO enrichment and depletion (using goseq)
sym_goseq.df1 = goseq(sym_pwf.1, gene2cat = sym_GO.list, method = "Wallenius", use_genes_without_cat = T) %>% mutate(contrast = 'Dominant_SS')
sym_goseq.df2 = goseq(sym_pwf.2, gene2cat = sym_GO.list, method = "Wallenius", use_genes_without_cat = T) %>% mutate(contrast = 'Dominant_WT10')
sym_goseq.df3 = goseq(sym_pwf.3, gene2cat = sym_GO.list, method = "Wallenius", use_genes_without_cat = T) %>% mutate(contrast = 'Absent_C21')
```

### Pass gene IDs to enriched/depleted GO terms
```{r}
#Pass IDs of DEGs to output from enrichment (gene column = gene IDs of numInCat)
sym_GO_DEG.df <- sym_GO.df %>% filter(gene %in% sym_ALL.DGE_assayed$gene)

GO_sym.df1 <- aggregate(gene ~ GO_ID, sym_GO_DEG.df, FUN = function(x) paste(unique(x), collapse = "; ")) %>% rename('category' = 1)
GO_sym.df2 <- aggregate(gene ~ GO_ID, sym_GO_DEG.df, FUN = function(x) paste(unique(x), collapse = "; ")) %>% rename('category' = 1)
GO_sym.df3 <- aggregate(gene ~ GO_ID, sym_GO_DEG.df, FUN = function(x) paste(unique(x), collapse = "; ")) %>% rename('category' = 1)

sym_goseq.df1 <- left_join(sym_goseq.df1, GO_sym.df1, by = 'category')
sym_goseq.df2 <- left_join(sym_goseq.df2, GO_sym.df2, by = 'category')
sym_goseq.df3 <- left_join(sym_goseq.df3, GO_sym.df3, by = 'category')
```

### Keep significant GO terms (i.e., unadjusted pval < 0.05 & numDEInCat >=5), check  GO IDs for missing annotatations
```{r}
sym_enriched_signif.1 <- sym_goseq.df1 %>%
  mutate(padj_BH = p.adjust(over_represented_pvalue, method="BH")) %>%
  filter(over_represented_pvalue < 0.05) %>% 
  filter(numDEInCat >= 5) %>%
  mutate(signif = case_when(over_represented_pvalue & padj_BH < 0.05 ~ 'yes',
                            TRUE ~ 'no')) %>%
  mutate(pval = over_represented_pvalue, direction = 'enriched') %>%
  mutate(term = case_when(str_detect(category, 'GO:0018298') ~ 'obsolete protein-chromophore linkage',
                          str_detect(category, 'GO:1990939') ~ 'microtubule motor activity',
                          TRUE ~ term)) %>%
  mutate(ontology = case_when(str_detect(category, 'GO:0018298') ~ 'BP',
                              str_detect(category, 'GO:1990939') ~ 'MF',
                              TRUE ~ ontology))

sym_enriched_signif.2 <- sym_goseq.df2 %>%
  mutate(padj_BH = p.adjust(over_represented_pvalue, method="BH")) %>%
  filter(over_represented_pvalue < 0.05) %>%
  filter(numDEInCat >= 5) %>%
  mutate(signif = case_when(over_represented_pvalue & padj_BH < 0.05 ~ 'yes',
                            TRUE ~ 'no')) %>%
  mutate(pval = over_represented_pvalue, direction = 'enriched') %>%
  mutate(term = case_when(str_detect(category, 'GO:0018298') ~ 'obsolete protein-chromophore linkage',
                          str_detect(category, 'GO:0019012') ~ 'virion component',
                          str_detect(category, 'GO:0015696') ~ 'ammonium transmembrane transport',
                          str_detect(category, 'GO:0009405') ~ 'obsolete pathogenesis',
                          TRUE ~ term)) %>%
  mutate(ontology = case_when(str_detect(category, 'GO:0018298') ~ 'BP',
                              str_detect(category, 'GO:0019012') ~ 'CC',
                              str_detect(category, 'GO:0015696') ~ 'BP',
                              str_detect(category, 'GO:0009405') ~ 'BP',
                              TRUE ~ ontology))

sym_enriched_signif.3 <- sym_goseq.df3 %>%
  mutate(padj_BH = p.adjust(over_represented_pvalue, method="BH")) %>%
  filter(over_represented_pvalue < 0.05) %>% 
  filter(numDEInCat >= 5) %>%
  mutate(signif = case_when(over_represented_pvalue & padj_BH < 0.05 ~ 'yes',
                            TRUE ~ 'no')) %>%
  mutate(pval = over_represented_pvalue, direction = 'enriched') %>%
  mutate(term = case_when(str_detect(category, 'GO:0018298') ~ 'obsolete protein-chromophore linkage',
                          str_detect(category, 'GO:0034613') ~ 'protein localization',
                          str_detect(category, 'GO:0004597') ~ 'peptidyl-aspartic acid 3-dioxygenase activity',
                          TRUE ~ term)) %>%
  mutate(ontology = case_when(str_detect(category, 'GO:0018298') ~ 'BP',
                              str_detect(category, 'GO:0034613') ~ 'BP',
                              str_detect(category, 'GO:0004597') ~ 'MF',
                              TRUE ~ ontology))

sym_depleted_signif.1 <- sym_goseq.df1 %>%
  mutate(padj_BH = p.adjust(under_represented_pvalue, method="BH")) %>%
  filter(under_represented_pvalue < 0.05) %>% 
  filter(numDEInCat >= 5) %>%
  mutate(signif = case_when(under_represented_pvalue & padj_BH < 0.05 ~ 'yes',
                            TRUE ~ 'no')) %>%
  mutate(pval = under_represented_pvalue, direction = 'depleted') %>%
  mutate(term = case_when(str_detect(category, 'GO:0106311') ~ 'protein serine/threonine kinase activity',
                          TRUE ~ term)) %>%
  mutate(ontology = case_when(str_detect(category, 'GO:0106311') ~ 'MF',
                              TRUE ~ ontology))

sym_depleted_signif.2 <- sym_goseq.df2 %>%
  mutate(padj_BH = p.adjust(under_represented_pvalue, method="BH")) %>%
  filter(under_represented_pvalue < 0.05) %>% 
  filter(numDEInCat >= 5) %>%
  mutate(signif = case_when(under_represented_pvalue & padj_BH < 0.05 ~ 'yes',
                            TRUE ~ 'no')) %>%
  mutate(pval = under_represented_pvalue, direction = 'depleted') %>%
  mutate(term = case_when(str_detect(category, 'GO:0106311') ~ 'protein serine/threonine kinase activity',
                          TRUE ~ term)) %>%
  mutate(ontology = case_when(str_detect(category, 'GO:0106311') ~ 'MF',
                              TRUE ~ ontology))

sym_depleted_signif.3 <- sym_goseq.df3 %>% 
  mutate(padj_BH = p.adjust(under_represented_pvalue, method="BH")) %>%
  filter(under_represented_pvalue < 0.05) %>% 
  filter(numDEInCat >= 5) %>%
  mutate(signif = case_when(under_represented_pvalue & padj_BH < 0.05 ~ 'yes',
                            TRUE ~ 'no')) %>%
  mutate(pval = under_represented_pvalue, direction = 'depleted') %>%
  mutate(term = case_when(str_detect(category, 'GO:0018298') ~ 'obsolete protein-chromophore linkage',
                          str_detect(category, 'GO:0070615') ~ 'ATP-dependent chromatin remodeler activity',
                          TRUE ~ term)) %>%
  mutate(ontology = case_when(str_detect(category, 'GO:0018298') ~ 'MF',
                                str_detect(category, 'GO:0070615') ~ 'MF',
                                TRUE ~ ontology))

sym_GO_signif <- rbind(sym_enriched_signif.1, sym_enriched_signif.2) %>%
  rbind(sym_enriched_signif.3) %>%
  rbind(sym_depleted_signif.1) %>%
  rbind(sym_depleted_signif.2) %>%
  rbind(sym_depleted_signif.3) %>%
  dplyr::select(!c(over_represented_pvalue, under_represented_pvalue)) %>%
  filter(!(term %>% str_detect('obsolete')))
```

### Pass ORA results to df and repeat with other direction
```{r}
#sym_GO_signif_down <- sym_GO_signif
sym_GO_signif_down2 <- sym_GO_signif_down %>%
  mutate(direction2 = 'down') %>%
  filter(!(term %>%str_detect("cardiac|heart|kidney|neuro")))

#sym_GO_signif_up <- sym_GO_signif
sym_GO_signif_up2 <- sym_GO_signif_up %>%
  mutate(direction2 = 'up') %>%
  mutate(term = case_when(str_detect(category,'GO:0004365') ~'GAPDH (NAD+) (phosphorlyating) activity', #Abbreviate term
                          TRUE ~ term)) %>%
  filter(!(term %>%str_detect('axon|cardiac|costamere|node')))

sym_GO_analysis.export <- rbind(sym_GO_signif_up2, sym_GO_signif_down2) %>%
  dplyr::select(contrast, direction2, direction, category, numDEInCat, numInCat, category,
                ontology, term, pval, padj_BH, gene) %>%
  rename('Photosymbiont lineage' = contrast,
         'Expression direction' = direction2,
         'Enriched/depleted' = direction,
         'GO ID' = category,
         'GO aspect' = ontology,
         'GO term' = term,
         'p-value' = pval,
         'p-value (adj)' = padj_BH,
         'DEGs' = gene) %>%
   write_csv(path = "./Table S5.csv", col_names = TRUE)
```

### Visualise results from ORA - upregulated DEGs
```{r}
sym_GO_signif_up2$contrast <- factor(sym_GO_signif_up2$contrast,
                                     levels = c("Dominant_SS", "Dominant_WT10", "Absent_C21"))

#Plots
sym_enriched_BP.plot_up <- sym_GO_signif_up2 %>%
  filter(ontology == 'BP') %>%
  filter(direction == 'enriched') %>%
  #group_by(contrast) %>%
  #top_n(10, wt = -pval) %>% 
  mutate(gene_ratio = numDEInCat/numInCat) %>%
  ggplot(aes(x = gene_ratio, y = fct_reorder(term, gene_ratio), colour = contrast, size = numDEInCat, alpha = pval)) +
  geom_point(aes(shape = signif)) +
  scale_x_continuous('Proportion of DEGs in GO term', limits = c(0, 1), expand = c(0,0)) +
  scale_colour_manual('Photosymbiont lineage', 
                      values = c("#DB8941", "#9f6671", "#0e3b5c"),
                      labels = c('C. proliferum (SS)', 'C. proliferum (WT)', 'C21')) +
  scale_size_continuous('# DEGs in GO term', breaks = 1:125*25, limits = c(0, 125)) +
  scale_shape_manual(values = c(16, 17)) +
  scale_alpha_continuous('p-value', range = c(1, 0.25), breaks = c(0.001, 0.01, 0.02, 0.03, 0.04), 
                         labels = c('<0.001', '0.01', '0.02', '0.03', '0.04')) +
  labs(y = "Biological Processes") +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        plot.margin = unit(c(0,0,0,0), "null"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  guides(shape = F,
         colour = guide_legend(override.aes = list(size = 3)),
         alpha = guide_legend(override.aes = list(size = 3)))
sym_enriched_BP.plot_up

sym_enriched_MF.plot_up <- sym_GO_signif_up2 %>%
  filter(ontology == 'MF') %>%
  filter(direction == 'enriched') %>%
  #group_by(contrast) %>%
  #top_n(10, wt = -pval) %>% 
  mutate(gene_ratio = numDEInCat/numInCat) %>%
  ggplot(aes(x = gene_ratio, y = fct_reorder(term, gene_ratio), colour = contrast, size = numDEInCat, alpha = pval)) +
  geom_point(aes(shape = signif)) +
  scale_x_continuous('Proportion of DEGs in GO term', limits = c(0, 1), expand = c(0,0)) +
  scale_colour_manual('Photosymbiont lineage', 
                      values = c("#DB8941", "#9f6671", "#0e3b5c"),
                      labels = c('C. proliferum (SS)', 'C. proliferum (WT)', 'C21')) +
  scale_shape_manual(values = c(16, 17)) +
  scale_size_continuous('# DEGs in GO term', breaks = 1:125*25, limits = c(0, 125)) +
  scale_alpha_continuous('p-value', range = c(1, 0.25), breaks = c(0.001, 0.01, 0.02, 0.03, 0.04), 
                         labels = c('<0.001', '0.01', '0.02', '0.03', '0.04')) +
  labs(y = "Molecular Functions") +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        plot.margin = unit(c(0,0,0,0), "null"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  guides(shape = F,
         colour = guide_legend(override.aes = list(size = 3)),
         alpha = guide_legend(override.aes = list(size = 3)))
sym_enriched_MF.plot_up

sym_enriched_CC.plot_up <- sym_GO_signif_up2 %>%
  filter(ontology == 'CC') %>%
  filter(direction == 'enriched') %>%
  #group_by(contrast) %>%
  #top_n(10, wt = -pval) %>% 
  mutate(gene_ratio = numDEInCat/numInCat) %>%
  ggplot(aes(x = gene_ratio, y = fct_reorder(term, gene_ratio), colour = contrast, size = numDEInCat, alpha = pval)) +
  geom_point(aes(shape = signif)) +
  scale_x_continuous('Proportion of DEGs in GO term', limits = c(0, 1), expand = c(0,0)) +
  scale_colour_manual('Photosymbiont lineage', 
                      values = c("#DB8941", "#9f6671", "#0e3b5c"),
                      labels = c('C. proliferum (SS)', 'C. proliferum (WT)', 'C21')) +
  scale_shape_manual(values = c(16, 17)) +
  scale_size_continuous('# DEGs in GO term', breaks = 1:125*25, limits = c(0, 125)) +
  scale_alpha_continuous('p-value', range = c(1, 0.25), breaks = c(0.001, 0.01, 0.02, 0.03, 0.04), 
                         labels = c('<0.001', '0.01', '0.02', '0.03', '0.04')) +
  labs(y = "Cellular Components") +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        plot.margin = unit(c(0,0,0,0), "null"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  guides(shape = F,
         colour = guide_legend(override.aes = list(size = 3)),
         alpha = guide_legend(override.aes = list(size = 3)))
sym_enriched_CC.plot_up

sym_ORA.plots_up <- ggarrange(sym_enriched_BP.plot_up + rremove('xlab') + rremove('x.text'),
                           sym_enriched_MF.plot_up + rremove('xlab') + rremove('x.text'),
                           sym_enriched_CC.plot_up,
                           common.legend = T,
                           legend = 'right',
                           heights = c(1, 1, 0.5),
                           widths = c(1, 1, 1),
                           align = 'hv',
                           ncol = 1)
sym_ORA.plots_up

Fig.S4 <- sym_ORA.plots_up
```

### Visualise results from ORA - downgulated DEGs
```{r}
sym_GO_signif_down2$contrast <- factor(sym_GO_signif_down2$contrast,
                                       levels = c("Dominant_SS", "Dominant_WT10", "Absent_C21"))

#Plots
sym_enriched_BP.plot_down <- sym_GO_signif_down2 %>%
  filter(ontology == 'BP') %>%
  filter(direction == 'enriched') %>%
  #group_by(contrast) %>%
  #top_n(10, wt = -pval) %>% 
  mutate(gene_ratio = numDEInCat/numInCat) %>%
  ggplot(aes(x = gene_ratio, y = fct_reorder(term, gene_ratio), colour = contrast, size = numDEInCat, alpha = pval)) +
  geom_point(aes(shape = signif)) +
  scale_x_continuous('Proportion of DEGs in GO term', limits = c(0, 1.02), expand = c(0,0)) +
  scale_colour_manual('Photosymbiont lineage', 
                      values = c("#DB8941", "#9f6671", "#0e3b5c"),
                      labels = c('C. proliferum (SS)', 'C. proliferum (WT)', 'C21')) +
  scale_shape_manual(values = c(16, 17)) +
  scale_size_continuous('# DEGs in GO term', breaks = 1:400*75, limits = c(0, 400)) +
  scale_alpha_continuous('p-value', range = c(1, 0.25), breaks = c(0.001, 0.01, 0.02, 0.03, 0.04), 
                         labels = c('<0.001', '0.01', '0.02', '0.03', '0.04')) +
  labs(y = "Biological Processes") +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        plot.margin = unit(c(0,0,0,0), "null"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  guides(shape = F, 
         colour = guide_legend(override.aes = list(size = 3)),
         alpha = guide_legend(override.aes = list(size = 3)))
sym_enriched_BP.plot_down

sym_enriched_MF.plot_down <- sym_GO_signif_down2 %>%
  filter(ontology == 'MF') %>%
  filter(direction == 'enriched') %>%
  #group_by(contrast) %>%
  #top_n(10, wt = -pval) %>% 
  mutate(gene_ratio = numDEInCat/numInCat) %>%
  ggplot(aes(x = gene_ratio, y = fct_reorder(term, gene_ratio), colour = contrast, size = numDEInCat, alpha = pval)) +
  geom_point(aes(shape = signif)) +
  scale_x_continuous('Proportion of DEGs in GO term', limits = c(0, 1.02), expand = c(0,0)) +
  scale_colour_manual('Photosymbiont lineage', 
                      values = c("#DB8941", "#9f6671", "#0e3b5c"),
                      labels = c('C. proliferum (SS)', 'C. proliferum (WT)', 'C21')) +
  scale_shape_manual(values = c(16, 17)) +
  scale_size_continuous('# DEGs in GO term', breaks = 1:400*75, limits = c(0, 400)) +
  scale_alpha_continuous('p-value', range = c(1, 0.25), breaks = c(0.001, 0.01, 0.02, 0.03, 0.04), 
                         labels = c('<0.001', '0.01', '0.02', '0.03', '0.04')) +
  labs(y = "Molecular Functions") +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        plot.margin = unit(c(0,0,0,0), "null"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  guides(shape = F,
         colour = guide_legend(override.aes = list(size = 3)),
         alpha = guide_legend(override.aes = list(size = 3)))
sym_enriched_MF.plot_down

sym_enriched_CC.plot_down <- sym_GO_signif_down2 %>%
  filter(ontology == 'CC') %>%
  filter(direction == 'enriched') %>%
  #group_by(contrast) %>%
  #top_n(10, wt = -pval) %>% 
  mutate(gene_ratio = numDEInCat/numInCat) %>%
  ggplot(aes(x = gene_ratio, y = fct_reorder(term, gene_ratio), colour = contrast, size = numDEInCat, alpha = pval)) +
  geom_point(aes(shape = signif)) +
  scale_x_continuous('Proportion of DEGs in GO term', limits = c(0, 1.02), expand = c(0,0)) +
  scale_colour_manual('Photosymbiont lineage', 
                      values = c("#DB8941", "#9f6671", "#0e3b5c"),
                      labels = c('C. proliferum (SS)', 'C. proliferum (WT)', 'C21')) +
  scale_shape_manual(values = c(16, 17)) +
  scale_size_continuous('# DEGs in GO term', breaks = 1:400*75, limits = c(0, 400)) +
  scale_alpha_continuous('p-value', range = c(1, 0.25), breaks = c(0.001, 0.01, 0.02, 0.03, 0.04), 
                         labels = c('<0.001', '0.01', '0.02', '0.03', '0.04')) +
  labs(y = "Cellular Components") +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        plot.margin = unit(c(0,0,0,0), "null"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  guides(shape = F,
         colour = guide_legend(override.aes = list(size = 3)),
         alpha = guide_legend(override.aes = list(size = 3)))
sym_enriched_CC.plot_down

sym_ORA.plots_down <- ggarrange(sym_enriched_BP.plot_down + rremove('xlab') + rremove('x.text'),
                           sym_enriched_MF.plot_down + rremove('xlab') + rremove('x.text'),
                           sym_enriched_CC.plot_down,
                           common.legend = T,
                           legend = 'right',
                           heights = c(1, 1, 0.5),
                           widths = c(1, 1, 1),
                           align = 'hv',
                           ncol = 1)
sym_ORA.plots_down

Fig.S5 <- sym_ORA.plots_down
```

### Export ORA plots 
```{r}
ggsave(plot = Fig.S4, "./Fig. S4.pdf",
       path = "./",
       width = 15,
       height = 15,
       units = 'in',
       dpi = 300)

ggsave(plot = Fig.S5, "./Fig. S5.pdf",
       path = "./",
       width = 15,
       height = 12.5,
       units = 'in',
       dpi = 300)
```

### REVIGO analysis
```{r}
#Seperate enriched and up-regulated GOs according to photosymbiont lineage
sym_GO_REVIGO_up <- sym_GO_signif_up2 %>%
  dplyr::select(category, contrast, pval) %>%
  rename('GO_id' = 1, 'p-value' = 3)

sym_GO_REVIGO_up_SS <- sym_GO_REVIGO_up %>%
  filter(contrast == 'Dominant_SS') %>%
  dplyr::select(-contrast) %>%
  write_csv('./Enriched & up-regulated GOs - SS.csv')

sym_GO_REVIGO_up_WT10 <- sym_GO_REVIGO_up %>%
  filter(contrast == 'Dominant_WT10') %>%
  dplyr::select(-contrast) %>%
  write_csv('./Enriched & up-regulated GOs - WT10.csv')

sym_GO_REVIGO_up_C21 <- sym_GO_REVIGO_up %>%
  filter(contrast == 'Absent_C21') %>%
  dplyr::select(-contrast) %>%
  write_csv('./Enriched & up-regulated GOs - C21.csv')

#Repeat for doan-regulated GOs
sym_GO_REVIGO_down <- sym_GO_signif_down2 %>%
  dplyr::select(category, contrast, pval)

sym_GO_REVIGO_down_SS <- sym_GO_REVIGO_down %>%
  filter(contrast == 'Dominant_SS') %>%
  dplyr::select(-contrast) %>%
  write_csv('./Enriched & down-regulated GOs - SS.csv')

sym_GO_REVIGO_down_WT10 <- sym_GO_REVIGO_down %>%
  filter(contrast == 'Dominant_WT10') %>%
  dplyr::select(-contrast) %>%
  write_csv('./Enriched & down-regulated GOs - WT10.csv')

sym_GO_REVIGO_down_C21 <- sym_GO_REVIGO_down %>%
  filter(contrast == 'Absent_C21') %>%
  dplyr::select(-contrast) %>%
  write_csv('./Enriched & down-regulated GOs - C21.csv')
```

### Identify significant DEGs associated with enriched GO terms (focus on erniched GO terms)
```{r}
sym_GO_signif_DEGs <- rbind(sym_GO_signif_up2, sym_GO_signif_down2) %>%
  filter(direction == 'enriched') %>%
  dplyr::select(category, term, ontology, contrast, direction2, gene) %>%
  separate_rows(gene, sep = '; ')

sym_GO_signif_DEGs_count <- rbind(sym_GO_signif_up2, sym_GO_signif_down2) %>%
  filter(direction == 'enriched') %>%
  dplyr::select(category, term, ontology, contrast, gene, pval, padj_BH) %>%
  group_by(contrast) %>%
  filter(pval < 0.05) %>%
  dplyr::select(category, contrast) %>%
  distinct() %>%
  dplyr::count()

sym.DEGs2 <- sym.DEGs %>%
  group_by(contrast) %>%
  filter(padj < 0.05) %>%
  dplyr::select(contrast, gene, log2FoldChange, padj)
  
sym_GO_signif_DEGs2 <- left_join(sym_GO_signif_DEGs, sym.DEGs2,  by = c('contrast', 'gene')) %>%
  drop_na() %>%
  left_join(., dplyr::select(cprol_gene.annot, c(gene, protein)))
```

### Analysis of gene expression levels involved in GO terms of interest

#### Categorise DEGs associated with GO terms
```{r}
sym_GO_expression_raw <- sym_GO_signif_DEGs2 %>%
  unite("GO_gene", c('category', 'gene'), sep= "_", remove = F) %>%
  group_by(contrast) %>%
  mutate(group = case_when(
    str_detect(term,"[A|a]scorbate|[C|c]atalase|[G|g]lutathione|[P|p]eroxidase|[P|p]eroxiredoxin|[S|s]uperoxide") ~ 'ROS scavengers',
    str_detect(term,"[C|c]haperone|[D|d]naj|[H|h]eat shock|[H|h]sp70|[H|h]sp90") ~ 'Molecular chaperones',
    str_detect(term,"[C|c]hlorophyll|[C|c]hloroplast|[L|l]ight|[P|p]hotosystem|[P|p]hotosynthesis|[T|t]hylakoid") ~ 'Photosynthesis',
    str_detect(term,"[F|f]ructose|GAPDH|[G|g]lyceraldehyde|[G|g]lycolytic|[G|g]lucose|[P|p]hosphoglycerate") ~ 'Glycolysis',
    str_detect(term,"[L|l]ipid") ~ 'Lipid processes',
    str_detect(term,"[C|c]arbohydrate") ~ 'Carbohydrate transport',
    str_detect(term, "[A|a]mmonium|[N|n]itr[a|i]te") ~ 'Nitrogen transport',
    str_detect(term, "[O|o]xidative|[R|r]epair|[R|r]efolding|[U|u]nfolded") ~ 'Stress responses',
    TRUE ~ 'Other processes'))

sym_GO_expression_raw_count <- sym_GO_expression_raw %>%
  #filter(ontology != 'CC') %>%
  filter(group == 'ROS scavengers') %>%
  dplyr::select(category, contrast, term) %>%
  distinct() %>%
  group_by(contrast) %>%
  dplyr::count()
```

#### Volcano plot
```{r}
sym_GO_expression_raw.df1 <- sym_GO_expression_raw %>% 
  filter(contrast == 'Dominant_SS') %>%
  column_to_rownames(var = "GO_gene")

sym_GO_expression_raw.df2 <- sym_GO_expression_raw %>% 
  filter(contrast == 'Dominant_WT10') %>%
  column_to_rownames(var = "GO_gene")

sym_GO_expression_raw.df3 <- sym_GO_expression_raw %>% 
  filter(contrast == 'Absent_C21') %>%
  column_to_rownames(var = "GO_gene")

#Define different groups that will be coloured and shaped differently
group1.col <- c('Dominant_SS')
group2.col <- c('Dominant_WT10')
group3.col <- c('Absent_C21')

group1.shape <- c('ROS scavenging')
group2.shape <- c('Heat shock proteins')
group3.shape <- c('Photosynthesis')
group4.shape <- c('Nitrogen transport')
group5.shape <- c('Other processes')

#Create custom key-value pairs for different GO terms and photosymbiont groups
keyvals.col1 <- ifelse(
  sym_GO_expression_raw.df1$contrast %in% group1.col, '#DB8941',
  'black')
keyvals.col2 <- ifelse(
  sym_GO_expression_raw.df2$contrast %in% group2.col, '#FABB5A',
  'black')
keyvals.col3 <- ifelse(
  sym_GO_expression_raw.df3$contrast %in% group3.col, '#0e3b5c',
  'black')

keyvals.shape1 <- ifelse(
  sym_GO_expression_raw.df1$group %in% group1.shape, 16,
    ifelse(sym_GO_expression_raw.df1$group %in% group2.shape, 15,
           ifelse(sym_GO_expression_raw.df1$group %in% group3.shape, 17,
                  ifelse(sym_GO_expression_raw.df1$group %in% group4.shape, 18,
                         ifelse(sym_GO_expression_raw.df1$group %in% group5.shape, 4,
                                0)))))

keyvals.shape2 <- ifelse(
  sym_GO_expression_raw.df2$group %in% group1.shape, 16,
    ifelse(sym_GO_expression_raw.df2$group %in% group2.shape, 15,
           ifelse(sym_GO_expression_raw.df2$group %in% group3.shape, 17,
                  ifelse(sym_GO_expression_raw.df2$group %in% group4.shape, 18,
                         ifelse(sym_GO_expression_raw.df2$group %in% group5.shape, 4,
                                0)))))

keyvals.shape3 <- ifelse(
  sym_GO_expression_raw.df3$group %in% group1.shape, 16,
    ifelse(sym_GO_expression_raw.df3$group %in% group2.shape, 15,
           ifelse(sym_GO_expression_raw.df3$group %in% group3.shape, 17,
                  ifelse(sym_GO_expression_raw.df3$group %in% group4.shape, 18,
                         ifelse(sym_GO_expression_raw.df3$group %in% group5.shape, 4,
                                0)))))

keyvals.col1[is.na(keyvals.col1)] <- 'black'
names(keyvals.col1)[keyvals.col1 == '#DB8941'] <- 'SS'

keyvals.col2[is.na(keyvals.col2)] <- 'black'
names(keyvals.col2)[keyvals.col2 == '#FABB5A'] <- 'WT10'

keyvals.col3[is.na(keyvals.col3)] <- 'black'
names(keyvals.col3)[keyvals.col3 == '#0e3b5c'] <- 'C21'

keyvals.shape1[is.na(keyvals.shape1)] <- 0
names(keyvals.shape1)[keyvals.shape1 == 16] <- 'ROS scavenging'
names(keyvals.shape1)[keyvals.shape1 == 15] <- 'Heat shock proteins'
names(keyvals.shape1)[keyvals.shape1 == 17] <- 'Photosynthesis'
names(keyvals.shape1)[keyvals.shape1 == 18] <- 'Nitrogen transport'
names(keyvals.shape1)[keyvals.shape1 == 4] <- 'Other processes'

keyvals.shape2[is.na(keyvals.shape2)] <- 0
names(keyvals.shape2)[keyvals.shape2 == 16] <- 'ROS scavenging'
names(keyvals.shape2)[keyvals.shape2 == 15] <- 'Heat shock proteins'
names(keyvals.shape2)[keyvals.shape2 == 17] <- 'Photosynthesis'
names(keyvals.shape2)[keyvals.shape2 == 18] <- 'Nitrogen transport'
names(keyvals.shape2)[keyvals.shape2 == 4] <- 'Other processes'

keyvals.shape3[is.na(keyvals.shape3)] <- 0
names(keyvals.shape3)[keyvals.shape3 == 16] <- 'ROS scavenging'
names(keyvals.shape3)[keyvals.shape3 == 15] <- 'Heat shock proteins'
names(keyvals.shape3)[keyvals.shape3 == 17] <- 'Photosynthesis'
names(keyvals.shape3)[keyvals.shape3 == 18] <- 'Nitrogen transport'
names(keyvals.shape3)[keyvals.shape3 == 4] <- 'Other processes'

sym_GO_expression_raw.volc_plot1 <- EnhancedVolcano(sym_GO_expression_raw.df1, lab = NA,
                                                   x = 'log2FoldChange', y = 'padj_BH',
                                                   xlim = c(-6, 6), #ylim = c(0, -log10(10e-52)),
                                                   title = '', subtitle = '', caption = '',
                                                   legendPosition = "top", legendLabSize = 8, legendIconSize = 2,
                                                   pCutoff = 0.05, FCcutoff = 0.58,
                                                   cutoffLineWidth = 0.25, cutoffLineCol = "black", 
                                                   cutoffLineType = "dashed",
                                                   colCustom = keyvals.col1,
                                                   shapeCustom = keyvals.shape1,
                                                   pointSize = c(ifelse(sym_GO_expression_raw.df1$group == 
                                                                          'Other processes', 0.5, 2)),
                                                   gridlines.minor = F, gridlines.major = F,
                                                   labSize = 3.0, axisLabSize = 14, colAlpha = 1) + 
  theme(plot.margin=unit(c(-2,-0.25,-0.95,0), "cm")) +
  guides(colour = F)

sym_GO_expression_raw.volc_plot2 <- EnhancedVolcano(sym_GO_expression_raw.df2, lab = NA,
                                                   x = 'log2FoldChange', y = 'padj_BH',
                                                   xlim = c(-6, 6), #ylim = c(0, -log10(10e-52)),
                                                   title = '', subtitle = '', caption = '',
                                                   legendPosition = "top", legendLabSize = 8, legendIconSize = 2,
                                                   pCutoff = 0.05, FCcutoff = 0.58,
                                                   cutoffLineWidth = 0.25, cutoffLineCol = "black", 
                                                   cutoffLineType = "dashed",
                                                   colCustom = keyvals.col2,
                                                   shapeCustom = keyvals.shape2,
                                                   pointSize = c(ifelse(sym_GO_expression_raw.df2$group == 
                                                                          'Other processes', 0.5, 2)),
                                                   gridlines.minor = F, gridlines.major = F,
                                                   labSize = 3.0, axisLabSize = 14, colAlpha = 1) + 
  theme(plot.margin=unit(c(-2,-0.25,-0.95,0), "cm")) +
  guides(colour = F)

sym_GO_expression_raw.volc_plot3 <- EnhancedVolcano(sym_GO_expression_raw.df3, lab = NA,
                                                   x = 'log2FoldChange', y = 'padj_BH',
                                                   xlim = c(-6, 6), #ylim = c(0, -log10(10e-52)),
                                                   title = '', subtitle = '', caption = '',
                                                   legendPosition = "top", legendLabSize = 8, legendIconSize = 2,
                                                   pCutoff = 0.05, FCcutoff = 0.58,
                                                   cutoffLineWidth = 0.25, cutoffLineCol = "black", 
                                                   cutoffLineType = "dashed",
                                                   colCustom = keyvals.col3,
                                                   shapeCustom = keyvals.shape3,
                                                   pointSize = c(ifelse(sym_GO_expression_raw.df3$group ==
                                                                          'Other processes', 0.5, 2)),
                                                   gridlines.minor = F, gridlines.major = F,
                                                   labSize = 3.0, axisLabSize = 14, colAlpha = 1) + 
  theme(plot.margin=unit(c(-2,-0.25,-0.95,0), "cm")) +
  guides(colour = F)

sym_GO_expression_raw.volc_plots <- ggarrange(sym_GO_expression_raw.volc_plot1 + rremove('ylab') + rremove('xlab'),
                                              sym_GO_expression_raw.volc_plot2 + rremove('ylab') + rremove('xlab'),
                                              sym_GO_expression_raw.volc_plot3 + rremove('ylab') + rremove('xlab'),
                                              common.legend = T,
                                              legend = "top",
                                              heights = c(1, 1, 1),
                                              align = 'hv',
                                              ncol = 3)
sym_GO_expression_raw.volc_plots

sym_GO_expression_raw.volc_plots_annot <- annotate_figure(sym_GO_expression_raw.volc_plots,
                                                          bottom = text_grob(bquote(bold(~Log[2]~"Fold Change")),
                                                                             color = "black", face = "bold", size = 14),
                                                          left = text_grob(bquote(bold(~-Log[10] ~ italic(P))),
                                                                           color= "black",face = "bold",size = 14,rot = 90))
sym_GO_expression_raw.volc_plots_annot
Fig.5B <- sym_GO_expression_raw.volc_plots_annot
```

#### Density plot of expression levels of enriched GO terms
```{r}
sym_GO_expression_raw$contrast <- factor(sym_GO_expression_raw$contrast, 
                                         levels = c("Dominant_SS", "Dominant_WT10", "Absent_C21"))

sym_GO_expression_raw.ridge_plot <- sym_GO_expression_raw %>%
  filter(ontology != 'CC') %>%
  filter(group != 'Other processes') %>%
  ggplot(aes(x = log2FoldChange, y = reorder(term, log2FoldChange), fill = contrast, colour = contrast)) +
  geom_vline(aes(xintercept = 0), linetype = 'dashed', size = 0.25) +
  geom_density_ridges(jittered_points = TRUE, position = position_points_jitter(height = 0),
                      point_shape = "|", point_size = 2, point_alpha = 1,
                      alpha = 0.5, scale = 1, size = 0) +
  scale_fill_manual('Photosymbiont lineage', 
                    labels = c('C. proliferum (SS)', 'C. proliferum (WT10)', 'C21'),
                    values = c('#DB8941', '#9f6671','#0e3b5c')) +
  scale_colour_manual('Photosymbiont lineage', 
                    labels = c('C. proliferum (SS)', 'C. proliferum (WT10)', 'C21'),
                    values = c('#DB8941', '#9f6671','#0e3b5c')) +
  scale_x_continuous(expand = c(0, 0), limits = c(-4, 4), breaks = -4:4*2) +
  facet_grid(group~., scales = "free_y", space="free_y") +
  theme_bw() +
  theme(panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank(),
        panel.border = element_blank(),
        strip.background = element_rect(color="white", fill="white", linewidth=0, linetype="solid"),
        strip.text.y = element_text(size = 12, face = "bold", angle = 0),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  labs(x = bquote(bold(~Log[2]~"Fold Change")), y = 'Gene ontology') +
  guides(colour = F)
sym_GO_expression_raw.ridge_plot

Fig.6B <- sym_GO_expression_raw.ridge_plot
```

#### Compare expression of HSP genes
```{r}
#Identify genes that encode for HSPs
sym_HSP_expr.df <- sym_GO_expression_raw %>%
  filter(str_detect(protein, 'Hsp|HSP|Heat shock|Chaperone protein [D|d]naJ|[D|d]naJ-like')) %>%
  mutate(group2 = case_when(
    str_detect(protein,"[H|h]eat shock 70 kDa protein|[H|h]sp70") ~ 'HSP70',
    str_detect(protein,"[H|h]eat shock 90 kDa protein|[H|h]eat shock protein 90|[H|h]sp90") ~ 'HSP90',
    str_detect(protein,"[D|d]naJ|[D|d]naJ-like") ~ 'HSP40/DnaJ',
    str_detect(protein,"[D|d]naJ|[D|d]naJ-like") ~ 'HSP40/DnaJ',
    TRUE ~ 'Other HSPs')) %>%
  filter(gene != 'evm.model.scf7180000353740.21') %>% #Gene is a Hop rather than a HSP
  mutate(contrast = as.factor(contrast)) %>%
  mutate(contrast = fct_relevel(contrast, "Dominant_SS", "Dominant_WT10", "Absent_C21")) %>%
  dplyr::select(contrast, gene, log2FoldChange, protein, group2) %>%
  distinct()

sym_HSP_expr.plot1 <- sym_HSP_expr.df %>%
  ggplot(aes(y = reorder(gene, log2FoldChange), x = log2FoldChange, fill = contrast)) +
  geom_vline(aes(xintercept = 0), linetype = 'dashed', size = 0.25) +
  #geom_point(pch = 21, size = 3) +
  geom_bar(stat = 'identity', position = position_dodge2(width = 1, preserve = "single")) +
  facet_grid(group2~., scales = "free_y", space="free_y") +
  scale_fill_manual('Photosymbiont strain', 
                      values = c("#DB8941", "#9f6671", "#0e3b5c"),
                      labels = c('C. proliferum (SS)', 'C. proliferum (WT)', 'C21')) +
  theme_bw() +
  theme(panel.grid.minor.x = element_blank(), 
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(colour = "black", size = 0.1), 
        panel.border = element_blank(),
        strip.text.y = element_text(size = 12, face = "bold"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  labs(x = bquote(bold(~Log[2]~"Fold Change")), y = 'Gene')
sym_HSP_expr.plot1

#Check if significant differences in HSP expression between photosymbiont strains and HSP type
tukey_hsd.HSP_strain <- aov(log2FoldChange ~ contrast, data = sym_HSP_expr.df) %>%
  tukey_hsd() %>%
  filter(p.adj < 0.05)

tukey_hsd.HSP_type <- aov(log2FoldChange ~ group2, data = sym_HSP_expr.df) %>%
  tukey_hsd() %>%
  filter(p.adj < 0.05)

sym_HSP_expr.plot2 <- sym_HSP_expr.df %>%
  ggplot(aes(x = contrast, y = log2FoldChange, fill = contrast)) +
  geom_hline(aes(yintercept = 0), linetype = 'dashed', size = 0.25) +
  geom_boxplot() +
  scale_x_discrete(labels = c('SS', 'WT10', 'C21')) +
  scale_fill_manual('Photosymbiont strain', 
                      values = c("#DB8941", "#9f6671", "#0e3b5c"),
                      labels = c('C. proliferum (SS)', 'C. proliferum (WT)', 'C21')) +
  theme_bw() +
  theme(panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank(),
        panel.border = element_blank(),
        strip.text.y = element_text(size = 12, face = "bold"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  labs(y = bquote(bold(~Log[2]~"Fold Change")), x = 'Photosymbiont strain') +
  guides(fill = F)
sym_HSP_expr.plot2

sym_HSP_expr.plot3 <- sym_HSP_expr.df %>%
  mutate(group2 = factor(group2, levels = c("HSP90", "HSP70", "HSP40/DnaJ", "Other HSPs"))) %>%
  ggplot() +
  geom_hline(aes(yintercept = 0), linetype = 'dashed', size = 0.25) +
  geom_boxplot(aes(x = group2, y = log2FoldChange), fill = 'grey') +
  stat_pvalue_manual(tukey_hsd.HSP_type, label = "p.adj.signif",  tip.length = 0,
                     y.position = 3) +
  scale_x_discrete() +
  theme_bw() +
  theme(panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank(),
        panel.border = element_blank(),
        strip.text.y = element_text(size = 12, face = "bold"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  labs(y = bquote(bold(~Log[2]~"Fold Change")), x = 'Heat shock protein category')
sym_HSP_expr.plot3

#VST gene counts at ambient and elevated temperature
sym_HSP_DEGs <- sym_GO_expression_raw %>% 
  filter(str_detect(protein, 'Hsp|HSP|Heat shock|Chaperone protein [D|d]naJ|[D|d]naJ-like')) %>%
  ungroup() %>% 
  select(gene) %>% 
  distinct()

sym_HSP_expr.plot4 <- sym.vsd_DEGs.mat2 %>%
  filter(gene %in% sym_HSP_DEGs$gene) %>%
  filter(Cprol_abundance == 'Dominant') %>%
  ggplot(aes(x = Cprol_strains, y = count_norm, fill = Temp_treatment)) +
  geom_boxplot() +
  #geom_jitter(width = 0.1, alpha = 0.5) +
  scale_colour_manual('Temperature treatment', values = c('#2171B5', '#CB181D')) +
  scale_fill_manual('Temperature treatment', values = c('#2171B5', '#CB181D')) +
  #scale_x_discrete('C. proliferum strain', expand = c(0,0)) +
  #scale_y_continuous('Gene z-score', expand = c(0,0)) +
  facet_grid(.~gene, scales = "free_y", space="free_y") +
  theme_bw() +
  theme(panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank(),
        panel.border = element_blank(),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black"))
sym_HSP_expr.plot4
```

#### Compare expression of glutathione genes
```{r}
#Identify genes that encode for glutathione processes
sym_glut_expr.df <- sym_GO_expression_raw %>%
  filter(str_detect(protein, 'Glutathione')) %>%
  mutate(group2 = case_when(
    str_detect(protein,"transferase|S-transferase") ~ 'GST',
    str_detect(protein,"reductase") ~ 'GR',
    str_detect(protein,"peroxidase") ~ 'GP',
    TRUE ~ 'Other glutathione processes')) %>%
  #filter(gene != 'evm.model.scf7180000353740.21') %>% #Gene is a Hop rather than a HSP
  mutate(contrast = as.factor(contrast)) %>%
  mutate(contrast = fct_relevel(contrast, "Dominant_SS", "Dominant_WT10", "Absent_C21")) %>%
  dplyr::select(contrast, gene, log2FoldChange, protein, group2) %>%
  distinct()

sym_glut_expr.plot <- sym_glut_expr.df %>%
  ggplot(aes(y = reorder(gene, log2FoldChange), x = log2FoldChange, fill = contrast)) +
  geom_vline(aes(xintercept = 0), linetype = 'dashed', size = 0.25) +
  #geom_point(pch = 21, size = 3) +
  geom_bar(stat = 'identity', position = position_dodge2(width = 1, preserve = "single")) +
  facet_grid(group2~., scales = "free_y", space="free_y") +
  scale_fill_manual('Photosymbiont strain', 
                      values = c("#DB8941", "#9f6671", "#0e3b5c"),
                      labels = c('C. proliferum (SS)', 'C. proliferum (WT)', 'C21')) +
  theme_bw() +
  theme(panel.grid.minor.x = element_blank(), 
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(colour = "black", size = 0.1), 
        panel.border = element_blank(),
        strip.text.y = element_text(size = 12, face = "bold"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  labs(x = bquote(bold(~Log[2]~"Fold Change")), y = 'Gene')
sym_glut_expr.plot

Fig.S6 <- sym_glut_expr.plot
```

#### Compare expression of nitrogen genes
```{r}
#Identify genes that encode for glutathione processes
sym_nitr_expr.df <- sym_GO_expression_raw %>%
  filter(str_detect(protein, 'High affinity nitrate transporter 2.5|Nitr[a|i]te reductase')) %>%
  mutate(group2 = case_when(
    str_detect(protein,"transporter") ~ 'NRT 2.5',
    str_detect(protein,"Nitrate reductase") ~ 'NR',
    str_detect(protein,"Nitrite reductase") ~ 'NIR',
    TRUE ~ 'Other nitrogen processes')) %>%
  mutate(contrast = as.factor(contrast)) %>%
  mutate(contrast = fct_relevel(contrast, "Dominant_SS", "Dominant_WT10", "Absent_C21")) %>%
  dplyr::select(contrast, gene, log2FoldChange, protein, group2) %>%
  distinct()

sym_nitr_expr.plot <- sym_nitr_expr.df %>%
  ggplot(aes(y = reorder(gene, log2FoldChange), x = log2FoldChange, fill = contrast)) +
  geom_vline(aes(xintercept = 0), linetype = 'dashed', size = 0.25) +
  #geom_point(pch = 21, size = 3) +
  geom_bar(stat = 'identity', position = position_dodge2(width = 1, preserve = "single")) +
  facet_grid(group2~., scales = "free_y", space="free_y") +
  scale_fill_manual('Photosymbiont strain', 
                      values = c("#DB8941", "#9f6671", "#0e3b5c"),
                      labels = c('C. proliferum (SS)', 'C. proliferum (WT)', 'C21')) +
  theme_bw() +
  theme(panel.grid.minor.x = element_blank(), 
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(colour = "black", size = 0.1), 
        panel.border = element_blank(),
        strip.text.y = element_text(size = 12, face = "bold"),
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 12),
        legend.position = 'top',
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(size = 14, colour = "black", face = 'bold'),
        axis.text.x = element_text(size = 12, colour = "black"), 
        axis.title.y = element_text(size = 14, colour = "black", face = 'bold'), 
        axis.text.y = element_text(size = 12, colour = "black")) +
  labs(x = bquote(bold(~Log[2]~"Fold Change")), y = 'Gene')
sym_nitr_expr.plot

Fig.S8 <- sym_nitr_expr.plot
```

#### Combine gene HSP expression plots
```{r}
sym_HSP.plot_comb <- ggarrange(sym_HSP_expr.plot2,
                               sym_HSP_expr.plot3,
                               labels = c("B", "C"),
                               align = "hv",
                               ncol = 2)
sym_HSP.plot_comb

Fig.S7 <- ggarrange(sym_HSP_expr.plot1,
                    sym_HSP.plot_comb,
                    labels = c("A", ""),
                    align = "hv",
                    heights = c(1, 0.5),
                    ncol = 1)
Fig.S7
```

#### Export ORA plots
```{r}
ggsave(plot = Fig.6B, "./Fig. 6.pdf",
       path = "./",
       width = 22.5,
       height = 12.5,
       units = 'in',
       dpi = 300)

ggsave(plot = Fig.S6, "./Fig. S6.pdf",
       path = "./",
       width = 10,
       height = 5,
       units = 'in',
       dpi = 300)

ggsave(plot = Fig.S7, "./Fig. S7.pdf",
       path = "./",
       width = 10,
       height = 10,
       units = 'in',
       dpi = 300)

ggsave(plot = Fig.S8, "./Fig. S8.pdf",
       path = "./",
       width = 10,
       height = 5,
       units = 'in',
       dpi = 300)
```

## GSEA analysis
```{r}
sym.DEGs.gsea_SS <- sym.DEGs %>%
  filter(contrast == 'Dominant_SS') %>%
  drop_na() %>%
  distinct() %>%
  dplyr::select(gene, log2FoldChange) %>%
  arrange(-log2FoldChange) %>%
  pull(log2FoldChange, gene)

sym.DEGs.gsea_WT10 <- sym.DEGs %>%
  filter(contrast == 'Dominant_WT10') %>%
  drop_na() %>%
  distinct() %>%
  dplyr::select(gene, log2FoldChange) %>%
  arrange(-log2FoldChange) %>%
  pull(log2FoldChange, gene)

sym.DEGs.gsea_C21 <- sym.DEGs %>%
  filter(contrast == 'Absent_C21') %>%
  drop_na() %>%
  distinct() %>%
  dplyr::select(gene, log2FoldChange) %>%
  arrange(-log2FoldChange) %>%
  pull(log2FoldChange, gene)

#Introduce some variation to avoid DEGs with same expression levels
tied_values_SS <- which(duplicated(sym.DEGs.gsea_SS) | duplicated(sym.DEGs.gsea_SS, fromLast = TRUE))
tied_values_WT10 <- which(duplicated(sym.DEGs.gsea_WT10) | duplicated(sym.DEGs.gsea_WT10, fromLast = TRUE))
tied_values_C21 <- which(duplicated(sym.DEGs.gsea_C21) | duplicated(sym.DEGs.gsea_C21, fromLast = TRUE))

sym.DEGs.gsea_SS[tied_values_SS] <- sym.DEGs.gsea_SS[tied_values_SS] + runif(length(tied_values_SS), min = 0, max = 0.001)
sym.DEGs.gsea_WT10[tied_values_WT10] <- sym.DEGs.gsea_WT10[tied_values_WT10] + runif(length(tied_values_WT10), min = 0, max = 0.001)
sym.DEGs.gsea_C21[tied_values_C21] <- sym.DEGs.gsea_C21[tied_values_C21] + runif(length(tied_values_C21), min = 0, max = 0.001)

#Preapre GO list
sym_GO.list2 <- split(sym_GO.df$gene, sym_GO.df$GO_ID)

#Run GSEA analysis
fgseaRes_SS <- fgsea(pathways = sym_GO.list2, stats = sym.DEGs.gsea_SS,
                  eps = 0.0, minSize  = 5, maxSize  = 500) %>%
  as_tibble() %>% 
  arrange(padj)

fgseaRes_WT10 <- fgsea(pathways = sym_GO.list2, stats = sym.DEGs.gsea_WT10,
                  eps = 0.0, minSize  = 5, maxSize  = 500) %>%
  as_tibble() %>% 
  arrange(padj)

fgseaRes_C21 <- fgsea(pathways = sym_GO.list2, stats = sym.DEGs.gsea_C21,
                  eps = 0.0, minSize  = 5, maxSize  = 500) %>%
  as_tibble() %>% 
  arrange(padj)
```
